---
title: "Exercises 4"
author: "Alexandria Hughes"
date: "12/11/2017"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

Here is a [link](https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-4.Rmd) to the text of these exercises.

# Question 1

What is the difference between `bed_shuffle()` and `bed_random()`? Explain your answer. No code is necessary, but you are welcome to show some example usage if it helps your explanation. 

## Interpretation



# Question 2

 Write a function that computes the final tibble needed for plotting a metagene. Explain in your interpretation why writing a function is useful for this task.
 
 The function should take three arguments.
  - `genes_bed` is a bed tibble that contains positions of human genes. (`chrom`, `start`, `end`, `name`, `score`, `strand`)
  - `chipseq_bedgraph` is a bedgraph tibble (`chrom`, `start`, `end`, `value`) 
  - `genome_bed` is a genome tibble (`chrom`, `size`)
  
Run your function using the following bed tibbles as input arguments

```{r files}
library(valr)
library(eda)
library(tidyverse)

# genes_bed 
genes_file <- system.file("extdata", "hg19_genes.bed.gz", package = "eda")
genes <- read_bed(genes_file, n_fields = 6)

# chipseq_bedgraph
bed <- read_bed(valr_example("hela.h3k4.chip.bg.gz"),
                n_fields = 4)
bg <- select(bed, everything(), value = name)
bg <- mutate(bg, 
              value = as.numeric(value))

# genome_bed 
genome_file <- system.file("extdata", "hg19_genome.txt.gz", package = "eda")
genome <- read_genome(genome_file)
```

Your output tibble should look like this:

```r
#> # A tibble: 201 x 2
#>    .win_id total_coverage
#>      <int>          <dbl>
#>  1       1       46.97059
#>  2       2       56.48529
#>  3       3       47.42667
#>  4       4       53.97297
#>  5       5       48.36842
#>  6       6       56.05333
#>  7       7       47.08861
#>  8       8       55.15190
#>  9       9       47.33750
#> 10      10       55.94937
#> # ... with 191 more rows
```


## Strategy

```{r q2}
compute_metagene <- function(chipseq_bedgraph,
                                genes_bed,
                              genome_bed){
  output <- genes %>% 
    filter(strand == "+") %>% 
    mutate(end = start + 1) %>% 
    bed_slop(genome, both = 1000) %>% 
    bed_makewindows(genome, win_size = 10) %>% 
    bed_map(., 
            bg,
            total_reads = sum(value)) %>% 
    group_by(.win_id) %>%
    summarize(total_coverage = mean(total_reads, na.rm = TRUE)) 
  output
}

metagene <- compute_metagene(bg, genes, genome)
metagene
```

## Interpretation

# Question 3

Write a function that plots a metagene. The function should take one argument, the output from question 2, and produce a plot. Run the function on the output of question 2, and show your plot. 

Here's some pseudo-code to get you started. I've also included some code that fixes the x-axis. 

``` r
plot_metagene <- function(dat) {
  
  region_size <- 1000
  win_size <- 10

  x_labels <- seq(-region_size, region_size, by = win_size * 25)
  x_breaks <- seq(1, 201, by = 25)

  ggplot(..., aes(..., ...)) +
  ... +
  scale_x_continuous(labels = x_labels, breaks = x_breaks)  +
}

plot_metagene(metagene_data)
```

## Strategy

```{r q3}
plot_metagene <- function(dat) {
  
  region_size <- 1000
  win_size <- 10

  x_labels <- seq(-region_size, region_size, by = win_size * 25)
  x_breaks <- seq(1, 201, by = 25)

  ggplot(dat, aes(.win_id, total_coverage)) +
  geom_point()+
  scale_x_continuous(labels = x_labels, breaks = x_breaks)
  
}

plot_metagene(metagene)
```

## Interpretation

# Question 4

Use the `brauer_gene_exp` table to answer the following questions.

+ Are there any rows with `NA` values in `brauer_gene_exp`? Show your work. Hint: calculate the number of rows before and after removing `NA` values.

+ Which `systematic_name` and `nutrient` combination has the highest expression value? Interpret this result in the section below, keeping in mind that the `nutrient` condition indicates when the nutrient is *omitted* from the growth medium. Look up this gene at the [Saccharomyces Genome Database](http://yeastgenome.org).

## Strategy

```{r q4}
#Rows with NAs: first make blank entries NAs, then count with and without NAs and subtract
brauer_gene_exp[brauer_gene_exp == ""] <- NA

all_rows <- nrow(brauer_gene_exp)
na_omit_rows <- na.omit(brauer_gene_exp) %>% nrow()
all_rows - na_omit_rows

#Combo of name/nutrient with highest exp value
brauer_gene_exp %>% group_by(systematic_name, nutrient) %>% summarize(mean_exp = mean(expression)) %>% arrange(desc(mean_exp))
```

## Interpretation

There are 47565 rows with NA values.
From the yeastgenome.org page, expression of YJR152W is constitutive but sensitive to nitrogen catabolite repression. Omitting ammonia withholds a nitrogen source, which may increase expression of it to compensate.