---
title: "Exercises 5"
author: "alexandria hughes"
date: "12/13/2017"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(eda)

update_eda()

library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(tidyverse)
library(cowplot)

```

```{r}
head(drug_resistant2_coldata)
```

# Overview

Here is a [link](https://raw.githubusercontent.com/rnabioco/eda/master/vignettes/exercises-5.Rmd) to the text of these exercises.

# Question 1

The `drug_resistant2_counts` and `drug_resistant2_coldata` objects contain a count matrix and metadata for an experiment in which two parental cell lines and their drug resistant derivatives were sequenced. You're interested in determining the changes in gene expression associated with acquired resistance to the drug. Build a DESeqDataSet from these objects and calculate differential gene expression. For now, ignore the contribution of cell line identity (`design = ~drug`). Convert the results to a tibble and use dyplr verbs to determine the number of genes that are differentially expressed between the parent and resistant group (padj <= 0.01).

## Strategy

```{r q1}

dds <- DESeqDataSetFromMatrix(countData = drug_resistant2_counts, colData = drug_resistant2_coldata, design = ~drug)
dds

```
```{r q1_part2}
nrow(dds)
# remove genes with little/ no exp
dds <- dds[rowSums(counts(dds)) > 1,]
nrow(dds)
```

```{r q1_part3}
dds <- DESeq(dds)
```
```{r q1_part4}
# Extract default results (No Drug vs. Drug A)
res <- results(dds)

# View results summary
summary(res)

# Convert dataframe to tbl
res_tibble <- as.data.frame(res) %>%
  tibble::rownames_to_column(var = "gene") %>%
  as.tibble()
res_tibble

#filter for differentially expressed genes, padj less/eq 0.01 -- number of rows is number of differentially exp genes
res_tibble %>% 
  filter(padj <= 0.01) %>% 
  nrow()
  
```


## Interpretation

There are 25 genes that are differentially expressed between the parent and resistant groups. I removed genes with little/no expression so I likely lost one of the genes that would have otherwise been counted (my neighbors in class got 26 for this).

# Question 2

The small number of differentially expressed genes in question 1 is unexpected in this experiment. Apply a regularized log transform to the DESeqDataSet from question 1 and generate a PCA plot to examine the data. What appears to be wrong with these data? Hint: it will be easier to interpret the PCA plot if the `intgroups` argument is a character vector containing both variables in colData.

## Strategy

```{r q2}
# rlog transform
rld <- rlog(dds,
            blind = FALSE)
# PCA plot 
plotPCA(rld, intgroup = c("drug", "cell_line")) +
  geom_text(aes(label = name))+
  scale_color_brewer(palette = "Set2")
```

## Interpretation

THe variance is most attributable to cell type instead of drug treatment.

# Question 3

There are two reasonable ways to deal with the problem identified in question 2. One involves subsetting the counts matrix and colData as in the pseudo-code below...
``` r
drug_resistant2_counts_fixed <- drug_resistant2_counts[,-c(?,?)]
drug_resistant2_coldata_fixed <- drug_resistant2_coldata[-c(?,?),]
```
...and the other requires only editing the colData...
``` r
drug_resistant2_coldata_fixed <- drug_resistant2_coldata
drug_resistant2_coldata_fixed$drug[?] <- "parent"
drug_resistant2_coldata_fixed$drug[?] <- "resistant"
```
Use the approach that you think is most appropriate and generate a new, corrected DESeqDataSet. Justify your choice briefly in the interpretation. Calculate differential gene expression and report the new number of genes differentially expressed between the parent and resistant group.

## Strategy

```{r q3}
drug_resistant2_coldata_fixed <- drug_resistant2_coldata

drug_resistant2_coldata_fixed$drug[12] <- "parent"
drug_resistant2_coldata_fixed$drug[7] <- "resistant"

tibble::rownames_to_column(drug_resistant2_coldata) %>% as.tibble()

drug_resistant2_coldata_fixed

dds <-DESeqDataSetFromMatrix(drug_resistant2_counts, drug_resistant2_coldata_fixed, design = ~drug)

dds <- DESeq(dds)

res <- results(dds)

res_tibble <- as.data.frame(res) %>% tibble::rownames_to_column("gene") %>% as.tibble()

res_tibble %>% filter(padj <= 0.01) %>% nrow()

```

## Interpretation

I chose to swap because only 2 need to be edited - it's more straightforward than subsetting the counts matrix. The new number of differential genes is 454.

# Question 4

In addition to the problem corrected in question 3, the PCA plot in question 2 revealed that the majority of variance between groups in this experiment is not due to drug resistance. What is the primary source of variance? Build the DESeqDataSet once more, this time including a term in the design (`design = ~? + drug`) that will correct for the major source of uninteresting variance in this experiment. As before, calculate differential gene expression and report the new number of genes differentially expressed between the parent and resistant group.

## Strategy

```{r q4}
dds <-DESeqDataSetFromMatrix(drug_resistant2_counts, drug_resistant2_coldata_fixed, design = ~cell_line + drug)

dds <- DESeq(dds)

res <- results(dds)

res_tibble <- as.data.frame(res) %>% tibble::rownames_to_column("gene") %>% as.tibble()

res_tibble %>% filter(padj <= 0.01) %>% nrow()
```

## Interpretation

The primary source of variation is cell type. The new output, correcting for cell type and drug, is 1950 genes.