{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ver. 02022020\n",
    "# Lesson 3.5 - Strings and things\n",
    "\n",
    "We have reached a stage where you have started doing things with python that would take too long to do by hand or by using excel. Today we will learn the final few functions to enable you to write code that can process large datasets. At the end of this module, you will be able to read a data file and plot the results, all within jupyter.\n",
    "\n",
    "Today's topics:\n",
    " - Strings\n",
    " - File Handling\n",
    " - Importing packages\n",
    " - Basic plotting\n",
    " \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strings\n",
    "\n",
    "In typical code where you read information from a file, you are going to read each line of a file as a string. So knowing python's powerful options for string manipulation will help you write code that reads files like human eyes.\n",
    "\n",
    "Refreshing our memories, string is just a set of characters enclosed by quotation marks. We can manipulate strings in many ways with python that is powerful and very useful in a variety of contexts. \n",
    "\n",
    "Refresher: Each character in a string has an index. Consider the string \"A Test String\":<br>\n",
    "\n",
    "Character&emsp;Index<br>\n",
    "\"A\"&emsp;&emsp;0<br>\n",
    "\" \"&emsp;&emsp;1<br>\n",
    "\"T\"&emsp;&emsp;2<br>\n",
    "\"E\"&emsp;&emsp;3<br>\n",
    "\"S\"&emsp;&emsp;4<br>\n",
    "\"T\"&emsp;&emsp;5<br>\n",
    "\" \"&emsp;&emsp;6<br>\n",
    "\"S\"&emsp;&emsp;7<br>\n",
    "\"t\"&emsp;&emsp;8<br>\n",
    "\"r\"&emsp;&emsp;&ensp;9<br>\n",
    "\"i\"&emsp;&emsp;10<br>\n",
    "\"n\"&emsp;&emsp;11<br>\n",
    "\"g\"&emsp;&emsp;12<br>\n",
    "\n",
    "### Let us do some string manipulation!\n",
    "\n",
    "Let's some biological problems to illustrate string manipulation in python:\n",
    "\n",
    " - How do I calculate the percentage of positively charged residues in a protein sequence?\n",
    " - How do I get the reverse complement of a DNA sequence?\n",
    " - How do I calculate the GC% in my **raw** sequencing reads?\n",
    " \n",
    "#### In the course of these examples, we will learn how to: <br>\n",
    " - Count the occurance of a substring\n",
    " - Length of a string\n",
    " - Split a string\n",
    " - Replace characters in a string\n",
    " - Find characters in a string\n",
    " - Remove specific characters from a stirng\n",
    " - Ask if the string has alphabets or numbers\n",
    " - Open a file and read it inside jupyter\n",
    " - Use loops and datastructures to process files\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do I calculate the percentage of positively charged residues in a protein sequence?\n",
    "\n",
    "Let us take my favorite protein sequence and write code to find percentage of positively charged residues. Protein/DNA sequence files usually are in Fasta format:<br>\n",
    "\"A sequence in FASTA format begins with a single-line description, followed by lines of sequence data. The description line (defline) is distinguished from the sequence data by a greater-than (\">\") symbol at the beginning.\" ([From NCBI](https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp))<br><br>\n",
    "In reality, you would write code to process a fasta file. We will learn later today how to open and read a file, but for now, we will just initialize a string as a shortcut:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">histone_H3\n",
      "MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n"
     ]
    }
   ],
   "source": [
    "fasta_file=\">histone_H3\\nMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTE\\\n",
    "LLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\"\n",
    "print(fasta_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### String parsing\n",
    "\n",
    "String parsing can be a powerful way to process **large** or **many** files in batch mode to extract information relevant to us, avoiding manual curation. \n",
    "\n",
    "In the example of `fasta_file`, there are two lines:\n",
    "\n",
    "- First line is the sequence name, which starts with \">\"\n",
    "- Second line is the sequence itself\n",
    "\n",
    "How do we extract information from this fasta entry? We will use `string.split(separator)`. Here, `fasta_file` is the string, and `\\n` is the separator (split by lines).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Two lines in the fasta file, let us separate the lines\n",
    "fasta_lines = fasta_file.split('\\n')\n",
    "type(fasta_lines)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick refresher: What kind of data type is `fasta_lines`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(fasta_lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['>histone_H3', 'MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA']\n"
     ]
    }
   ],
   "source": [
    "print(fasta_lines)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were two lines, each of which are an element in the list `fasta_lines`. <br>Now, which of the elements is the header and which of them is the sequence? \n",
    "<br><br><br>\n",
    "How do I separate the header and the sequence?<br><br>\n",
    "It is always useful to write out or think out the pseudocode, so you can solve the problem in your head first, before writing any code.<br><br>\n",
    "Pseudocode:\n",
    "- The header in a fasta starts with `>`. We can check for that with `startswith()`. \n",
    "- We will go through each element of the list `fasta_lines` and check for the `>` sign. \n",
    "- If the `>` sign is absent, that element is the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\t>histone_H3\n",
      "Seq:\t\tMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n"
     ]
    }
   ],
   "source": [
    "for i in fasta_lines: #for each line\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        seq_name = i\n",
    "    else:\n",
    "        seq = i\n",
    "print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\t\"+seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, we have taught python to recognize the sequence now! What did we want to do again? Right, find the percentage of positive residues in the sequence.\n",
    "\n",
    "Lysine (K), Arginine (R), and Histodine (H) are the positive residues. We use `string.count(pattern)` to count occurences of these residues. Here, `seq` is the `string` and `H`, `K`, and `R` form the `pattern`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\t>histone_H3\n",
      "Seq:\t\tMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
      "Count of Positive Residues:33\n"
     ]
    }
   ],
   "source": [
    "for i in fasta_lines: #for each line\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        seq_name = i\n",
    "    else:\n",
    "        seq = i\n",
    "        #Count the positive residues when we get to a sequence line\n",
    "        count = seq.count('K') + seq.count('R') + seq.count('H')\n",
    "        #print the results inside the loop\n",
    "        print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\t\"+seq+\"\\nCount of Positive Residues:\"+str(count))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wait, this \"count\" doesn't take into account the protein length. We can't compare any two proteins on this metric. This brings us to the important concept of **normalization** which we will do a lot with any data we deal with to enable comparisons. How will you **normalize** the *positivity* of a protein??\n",
    "<br><br><br>\n",
    "We can **normalize** to total protein length. We use `len(string)` to find length of strings.\n",
    "(Can you think of other ways to normalize positive residues of a protein to enable comparisons?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\t>histone_H3\n",
      "Seq:\t\tMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
      "% of Positive Residues:24.3\n"
     ]
    }
   ],
   "source": [
    "for i in fasta_lines: #for each line\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        seq_name = i\n",
    "    else:\n",
    "        seq = i\n",
    "        #Count the positive residues when we get to a sequence line\n",
    "        count = seq.count('K') + seq.count('R') + seq.count('H')\n",
    "        percent_positive = count*100/len(seq)\n",
    "        #print the results inside the loop\n",
    "        #Use round to limit number of decimals (output in FLOAT)\n",
    "        print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\t\"+seq+\"\\n% of Positive Residues:\"+str(round(percent_positive,1))) #the 1 is the number of dec places to round to"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so is 24 high or low or par for the course? We need to study more proteins! Let us add two \"housekeeping\" proteins: GAPDH and Actin, and an RNA binding protein, HuR. Again, we are using a shortcut for now and assigning the sequences to a `string`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">histone_H3\n",
      "MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
      ">Actin\n",
      "MDDDIAALVVDNGSGMCKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQSKRGILTLKYPIEHGIVTNWDDMEKIWHHTFYNELRVAPEEHPVLLTEAPLNPKANREKMTQIMFETFNTPAMYVAIQAVLSLYASGRTTGIVMDSGDGVTHTVPIYEGYALPHAILRLDLAGRDLTDYLMKILTERGYSFTTTAEREIVRDIKEKLCYVALDFEQEMATAASSSSLEKSYELPDGQVITIGNERFRCPEALFQPSFLGMESCGIHETTFNSIMKCDVDIRKDLYANTVLSGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWISKQEYDESGPSIVHRKCF\n",
      ">Gapdh\n",
      "MGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTVKAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFTTMEKAGAHLQGGAKRVIISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHAITATQKTVDGPSGKLWRDGRGALQNIIPASTGAAKAVGKVIPELNGKLTGMAFRVPTANVSVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAGIALNDHFVKLISWYDNEFGYSNRVVDLMAHMASKE\n",
      ">HuR\n",
      "MSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHSLGYGFVNYVTAKDAERAINTLNGLRLQSKTIKVSYARPSSEVIKDANLYISGLPRTMTQKDVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPITVKFAANPNQNKNVALLSQLYHSPARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNASSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAMAIASLNGYRLGDKILQVSFKTNKSHK\n"
     ]
    }
   ],
   "source": [
    "fasta_file=\">histone_H3\\nMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLR\\\n",
    "FQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\\n>Actin\\nMDDDIAALVVDNGSGMCKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMG\\\n",
    "QKDSYVGDEAQSKRGILTLKYPIEHGIVTNWDDMEKIWHHTFYNELRVAPEEHPVLLTEAPLNPKANREKMTQIMFETFNTPAMYVAIQAVLSLYASGRTTGIVMDSGDG\\\n",
    "VTHTVPIYEGYALPHAILRLDLAGRDLTDYLMKILTERGYSFTTTAEREIVRDIKEKLCYVALDFEQEMATAASSSSLEKSYELPDGQVITIGNERFRCPEALFQPSFLG\\\n",
    "MESCGIHETTFNSIMKCDVDIRKDLYANTVLSGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWISKQEYDESGPSIVHRKCF\\n>\\\n",
    "Gapdh\\nMGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTVKAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFT\\\n",
    "TMEKAGAHLQGGAKRVIISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHAITATQKTVDGPSGKLWRDGRGALQNIIPASTGA\\\n",
    "AKAVGKVIPELNGKLTGMAFRVPTANVSVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAGIALNDHFVKLISWYDNEFGYSNRV\\\n",
    "VDLMAHMASKE\\n>HuR\\nMSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHSLGYGFVNYVTAKDAERAINTLNGLRLQSKTIK\\\n",
    "VSYARPSSEVIKDANLYISGLPRTMTQKDVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPITVKFAANPNQNKNVALLSQLYHSP\\\n",
    "ARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNASSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAMAIASLNGYRLGDKI\\\n",
    "LQVSFKTNKSHK\"\n",
    "print(fasta_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is the great thing about our code: If we were counting by hand, **we** have to do three times the work. But , we can use the exact same code that we wrote before - no extra effort!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\t>histone_H3\n",
      "Seq:\t\tMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
      "% of Positive Residues:24.3\n",
      "Seq name:\t>Actin\n",
      "Seq:\t\tMDDDIAALVVDNGSGMCKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQSKRGILTLKYPIEHGIVTNWDDMEKIWHHTFYNELRVAPEEHPVLLTEAPLNPKANREKMTQIMFETFNTPAMYVAIQAVLSLYASGRTTGIVMDSGDGVTHTVPIYEGYALPHAILRLDLAGRDLTDYLMKILTERGYSFTTTAEREIVRDIKEKLCYVALDFEQEMATAASSSSLEKSYELPDGQVITIGNERFRCPEALFQPSFLGMESCGIHETTFNSIMKCDVDIRKDLYANTVLSGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWISKQEYDESGPSIVHRKCF\n",
      "% of Positive Residues:12.3\n",
      "Seq name:\t>Gapdh\n",
      "Seq:\t\tMGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTVKAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFTTMEKAGAHLQGGAKRVIISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHAITATQKTVDGPSGKLWRDGRGALQNIIPASTGAAKAVGKVIPELNGKLTGMAFRVPTANVSVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAGIALNDHFVKLISWYDNEFGYSNRVVDLMAHMASKE\n",
      "% of Positive Residues:13.7\n",
      "Seq name:\t>HuR\n",
      "Seq:\t\tMSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHSLGYGFVNYVTAKDAERAINTLNGLRLQSKTIKVSYARPSSEVIKDANLYISGLPRTMTQKDVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPITVKFAANPNQNKNVALLSQLYHSPARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNASSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAMAIASLNGYRLGDKILQVSFKTNKSHK\n",
      "% of Positive Residues:14.1\n"
     ]
    }
   ],
   "source": [
    "#First create a list of lines\n",
    "fasta_lines = fasta_file.split('\\n')\n",
    "#Now run the exact same loop!\n",
    "for i in fasta_lines: #for each line\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        seq_name = i\n",
    "    else:\n",
    "        seq = i\n",
    "        #Count the positive residues when we get to a sequence line\n",
    "        count = seq.count('K') + seq.count('R') + seq.count('H')\n",
    "        percent_positive = count*100/len(seq)\n",
    "        #print the results inside the loop\n",
    "        print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\t\"+seq+\"\\n% of Positive Residues:\"+str(round(percent_positive,1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What other questions could we ask with `count()` and `len()` commands in biological sequences?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Replace\n",
    "We don't want the `>` sign as part of the sequence name. How do we get rid of it? <br>\n",
    "\n",
    "- `replace()` is like search & replace in text editing programs\n",
    "- str.replace(character to be replaced, replacement, count)\n",
    "- if count not specified, all instances replaced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HuR\n"
     ]
    }
   ],
   "source": [
    "seq_name = seq_name.replace(\">\",'')\n",
    "print(seq_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find motifs??\n",
    "\n",
    "So, HuR binds the motif `AUUUUUA`. How do we ask if an RNA sequence has this motif?\n",
    "\n",
    "- `find()` method finds the first occurence of the specified value.\n",
    "- `str.find ( prefix [ , start [ , end ] ] )`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'AUGCAUGCAGCUAGCUAAACCCGGGUUAUAUAGAAUUUUUAAAGUAGUAGGGCCGAGUAGAUUUUUAAGAGAGAUAUUUUUAGAGAGAUAGAUGAGAGG'\n",
    "seq.find('AUUUUUA') #this is just the first motif occurrence!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What if there are more motifs? How to find all of the motifs in a sequence?\n",
    "(Googled it, and found the [answer](https://www.google.com/search?q=find+multiple+instances+python+string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Motif found at 34\n",
      "Motif found at 60\n",
      "Motif found at 75\n"
     ]
    }
   ],
   "source": [
    "index=0\n",
    "while index < len(seq):\n",
    "    index = seq.find('AUUUUUA', index)\n",
    "    if index == -1:\n",
    "        break #make sure you get out if you don't find one (because -1 is always less than seq length this would o'w be inf)\n",
    "    print('Motif found at', index)\n",
    "    index += len('AUUUUUA') #pushes past the current find, to the next"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### An everyday example of string manipulation\n",
    "\n",
    "How to find the reverse complement of a *DNA* sequence?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = \"AUGCCAUGUUGTC\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If, ahem, someone enters an *RNA* sequence, let us clean it up first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"U\" in seq: #if it is RNA, let us change it to DNA before it gets degraded\n",
    "    better_seq = seq.replace('U','T')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let us find the complement. We will take advantage of the fact that a string is iterable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGCCATGTTGTC\n",
      "In the process\t ['T']\n",
      "In the process\t ['T', 'A']\n",
      "In the process\t ['T', 'A', 'C']\n",
      "In the process\t ['T', 'A', 'C', 'G']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A', 'C']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A', 'C', 'A']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A', 'C', 'A', 'A']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A', 'C', 'A', 'A', 'C']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A', 'C', 'A', 'A', 'C', 'A']\n",
      "In the process\t ['T', 'A', 'C', 'G', 'G', 'T', 'A', 'C', 'A', 'A', 'C', 'A', 'G']\n",
      "\n",
      "\n",
      "Here is a sequence and its complement:\n",
      "\n",
      "5'\t ATGCCATGTTGTC \t3'\n",
      "3'\t TACGGTACAACAG \t5'\n"
     ]
    }
   ],
   "source": [
    "print(better_seq)\n",
    "\n",
    "#First, let us define a list that will contain the complementary sequence\n",
    "complement_seq=[]\n",
    "\n",
    "#Iterate through the sequence (better_seq): this will go base by base\n",
    "#and then add the complementary base to complement_seq\n",
    "\n",
    "for i in range(0,len(better_seq)):\n",
    "    if(better_seq[i] == 'A'):\n",
    "        complement_seq.append('T')\n",
    "    elif(better_seq[i] == 'T'):\n",
    "        complement_seq.append('A')\n",
    "    elif(better_seq[i] == 'G'):\n",
    "        complement_seq.append('C')\n",
    "    elif(better_seq[i] == 'C'):\n",
    "        complement_seq.append('G')\n",
    "    #OK, is the code working as expected??\n",
    "    print(\"In the process\\t\",complement_seq)\n",
    "\n",
    "print(\"\\n\\nHere is a sequence and its complement:\\n\")\n",
    "print(\"5'\\t\",better_seq,\"\\t3'\")\n",
    "complement_better_seq = ''.join(complement_seq)\n",
    "print(\"3'\\t\",complement_better_seq,\"\\t5'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ori_Seq\t\t\t5'\t AUGCCAUGUUGTC \t3'\n",
      "DNA Seq\t\t\t5'\t ATGCCATGTTGTC \t3'\n",
      "Complement\t\t5'\t TACGGTACAACAG \t3'\n",
      "\n",
      "\n",
      "DNA Seq\t\t\t5'\t ATGCCATGTTGTC \t3'\n",
      "Reverse Complement\t5'\t GACAACATGGCAT \t3'\n",
      "GACAACATGGCAT\n"
     ]
    }
   ],
   "source": [
    "print(\"Ori_Seq\\t\\t\\t5'\\t\",seq,\"\\t3'\")\n",
    "print(\"DNA Seq\\t\\t\\t5'\\t\",better_seq,\"\\t3'\")\n",
    "print(\"Complement\\t\\t5'\\t\",complement_better_seq,\"\\t3'\\n\\n\")\n",
    "\n",
    "#Now reverse complement:\n",
    "#rev_better_seq = complement_better_seq[::-1]\n",
    "rev_better_seq = \"\".join(list(reversed(complement_better_seq)))\n",
    "print(\"DNA Seq\\t\\t\\t5'\\t\",better_seq,\"\\t3'\")\n",
    "print(\"Reverse Complement\\t5'\\t\",rev_better_seq,\"\\t3'\")\n",
    "\n",
    "print(rev_better_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What kind of characters does my string have?\n",
    " - isalpha() - are all the characters alphabets?\n",
    " - isnumeric() - are all characters numeric?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = \"ATGCATCGA\"\n",
    "seq.isalpha()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq.isnumeric()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### One way to check it is only a DNA sequence\n",
    "\n",
    "Remove \"A\", \"T\", \"G\", and \"C\". Does anything remain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Sequence seems like DNA\n"
     ]
    }
   ],
   "source": [
    "#input_seq = \"ATAGAGCUUU\"\n",
    "input_seq = \"ATAGAGC\"\n",
    "tmp_seq = input_seq\n",
    "tmp_seq = tmp_seq.replace(\"A\",\"\").replace(\"T\",\"\").replace(\"G\",\"\").replace(\"C\",\"\")\n",
    "print(tmp_seq)\n",
    "if(len(tmp_seq)>0):\n",
    "    print(\"Non DNA alphabets detected\")\n",
    "else:\n",
    "    print(\"Sequence seems like DNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading fasta files\n",
    "\n",
    "In the example where we calculated percentage positive residues, we put in the sequences by hand. This will hardly happen day-to-day: usually, you get a file that you need to process with your code. Let us see how we will do the same operations we did before, but when we are given a file of protein sequences (with all its quirks).<br><br>\n",
    "To read a file in python, we use the command `open()`\n",
    "\n",
    "\n",
    "### open() command\n",
    " - Two inputs - filename, mode\n",
    " - Mode can be \n",
    "     - 'r' for read\n",
    "     - 'w' for write (overwrites existing file)\n",
    "     - 'a' for append (adds to existing file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '_io.TextIOWrapper'>\n"
     ]
    }
   ],
   "source": [
    "a_file = open('protein_sequences.fasta', 'r' )\n",
    "# \"a_file\" is an object in which the whole file is stored\n",
    "\n",
    "#What datatype is \"a_file\"?\n",
    "print(type(a_file))\n",
    "\n",
    "#A good habit is to close the file after opening it. Always have the close command written out\n",
    "#and add code between the open and close commands\n",
    "a_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In most scenarios, we will read the file line by line. Here is how we do that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">HuR\n",
      "\n",
      "MSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHS \n",
      "\n",
      "LGYGFVNYVTAKDAERAINTLNGLRLQSKTIKVSYARPSSEVIKDANLYISGLPRTMTQK \n",
      "\n",
      "DVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPIT \n",
      "\n",
      "VKFAANPNQNKNVALLSQLYHSPARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNA\n",
      "\n",
      "SSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAM \n",
      "\n",
      "AIASLNGYRLGDKILQVSFKTNKSHK \n",
      "\n",
      ">Histone_H3\n",
      "\n",
      "MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTE\n",
      "\n",
      "LLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTI\n",
      "\n",
      "MPKDIQLARRIRGERA\n",
      "\n",
      ">Gapdh\n",
      "\n",
      "MGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTV\n",
      "\n",
      "KAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFTTMEKAGAHLQGGAKRVI\n",
      "\n",
      "ISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHA\n",
      "\n",
      "ITATQKTVDGPSGKLWRDGRGALQNIIPASTGAAKAVGKVIPELNGKLTGMAFRVPTANV\n",
      "\n",
      "SVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAG\n",
      "\n",
      "IALNDHFVKLISWYDNEFGYSNRVVDLMAHMASKE\n",
      "\n",
      ">Actin\n",
      "\n",
      "MCEEETTALVCDNGSGLCKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQ\n",
      "\n",
      "SKRGILTLKYPIEHGIITNWDDMEKIWHHSFYNELRVAPEEHPTLLTEAPLNPKANREKM\n",
      "\n",
      "TQIMFETFNVPAMYVAIQAVLSLYASGRTTGIVLDSGDGVTHNVPIYEGYALPHAIMRLD\n",
      "\n",
      "LAGRDLTDYLMKILTERGYSFVTTAEREIVRDIKEKLCYVALDFENEMATAASSSSLEKS\n",
      "\n",
      "YELPDGQVITIGNERFRCPETLFQPSFIGMESAGIHETTYNSIMKCDIDIRKDLYANNVL\n",
      "\n",
      "SGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWISK\n",
      "\n",
      "PEYDEAGPSIVHRKCF\n",
      "\n",
      "â€¨\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#print every line\n",
    "a_file = open('protein_sequences.fasta', 'r' )\n",
    "for line in a_file:\n",
    "    print(line)\n",
    "a_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wow, we can iterate over `a_file` just like the list `fasta_lines`! Can we then use the same code with a few edits??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\t>HuR\n",
      "\n",
      "Seq:\t\tMSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHS \n",
      "\n",
      "% of Positive Residues:12.9\n",
      "Seq name:\t>HuR\n",
      "\n",
      "Seq:\t\tLGYGFVNYVTAKDAERAINTLNGLRLQSKTIKVSYARPSSEVIKDANLYISGLPRTMTQK \n",
      "\n",
      "% of Positive Residues:14.5\n",
      "Seq name:\t>HuR\n",
      "\n",
      "Seq:\t\tDVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPIT \n",
      "\n",
      "% of Positive Residues:14.5\n",
      "Seq name:\t>HuR\n",
      "\n",
      "Seq:\t\tVKFAANPNQNKNVALLSQLYHSPARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNA\n",
      "\n",
      "% of Positive Residues:16.4\n",
      "Seq name:\t>HuR\n",
      "\n",
      "Seq:\t\tSSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAM \n",
      "\n",
      "% of Positive Residues:6.5\n",
      "Seq name:\t>HuR\n",
      "\n",
      "Seq:\t\tAIASLNGYRLGDKILQVSFKTNKSHK \n",
      "\n",
      "% of Positive Residues:21.4\n",
      "Seq name:\t>Histone_H3\n",
      "\n",
      "Seq:\t\tMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTE\n",
      "\n",
      "% of Positive Residues:31.1\n",
      "Seq name:\t>Histone_H3\n",
      "\n",
      "Seq:\t\tLLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTI\n",
      "\n",
      "% of Positive Residues:14.8\n",
      "Seq name:\t>Histone_H3\n",
      "\n",
      "Seq:\t\tMPKDIQLARRIRGERA\n",
      "\n",
      "% of Positive Residues:29.4\n",
      "Seq name:\t>Gapdh\n",
      "\n",
      "Seq:\t\tMGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTV\n",
      "\n",
      "% of Positive Residues:14.8\n",
      "Seq name:\t>Gapdh\n",
      "\n",
      "Seq:\t\tKAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFTTMEKAGAHLQGGAKRVI\n",
      "\n",
      "% of Positive Residues:14.8\n",
      "Seq name:\t>Gapdh\n",
      "\n",
      "Seq:\t\tISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHA\n",
      "\n",
      "% of Positive Residues:9.8\n",
      "Seq name:\t>Gapdh\n",
      "\n",
      "Seq:\t\tITATQKTVDGPSGKLWRDGRGALQNIIPASTGAAKAVGKVIPELNGKLTGMAFRVPTANV\n",
      "\n",
      "% of Positive Residues:13.1\n",
      "Seq name:\t>Gapdh\n",
      "\n",
      "Seq:\t\tSVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAG\n",
      "\n",
      "% of Positive Residues:14.8\n",
      "Seq name:\t>Gapdh\n",
      "\n",
      "Seq:\t\tIALNDHFVKLISWYDNEFGYSNRVVDLMAHMASKE\n",
      "\n",
      "% of Positive Residues:13.9\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tMCEEETTALVCDNGSGLCKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQ\n",
      "\n",
      "% of Positive Residues:9.8\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tSKRGILTLKYPIEHGIITNWDDMEKIWHHSFYNELRVAPEEHPTLLTEAPLNPKANREKM\n",
      "\n",
      "% of Positive Residues:19.7\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tTQIMFETFNVPAMYVAIQAVLSLYASGRTTGIVLDSGDGVTHNVPIYEGYALPHAIMRLD\n",
      "\n",
      "% of Positive Residues:6.6\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tLAGRDLTDYLMKILTERGYSFVTTAEREIVRDIKEKLCYVALDFENEMATAASSSSLEKS\n",
      "\n",
      "% of Positive Residues:13.1\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tYELPDGQVITIGNERFRCPETLFQPSFIGMESAGIHETTYNSIMKCDIDIRKDLYANNVL\n",
      "\n",
      "% of Positive Residues:9.8\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tSGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWISK\n",
      "\n",
      "% of Positive Residues:11.5\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tPEYDEAGPSIVHRKCF\n",
      "\n",
      "% of Positive Residues:17.6\n",
      "Seq name:\t>Actin\n",
      "\n",
      "Seq:\t\tâ€¨\n",
      "\n",
      "% of Positive Residues:0.0\n"
     ]
    }
   ],
   "source": [
    "a_file = open('protein_sequences.fasta', 'r' )\n",
    "for i in a_file: #for each line\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        seq_name = i\n",
    "    else:\n",
    "        seq = i\n",
    "        #Count the positive residues when we get to a sequence line\n",
    "        count = seq.count('K') + seq.count('R') + seq.count('H')\n",
    "        percent_positive = count*100/len(seq)\n",
    "        #print the results inside the loop\n",
    "        print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\t\"+seq+\"\\n% of Positive Residues:\"+str(round(percent_positive,1)))\n",
    "        #Put \"ST\" and \"EN\" to denote start and end of string that we read in:\n",
    "        #print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\tST##\"+seq+\"##EN\\n% of Positive Residues:\"+str(round(percent_positive,1)))\n",
    "a_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### \"We're not in Kansas anymore\"\n",
    "\n",
    "Till now, we were dealing with nice examples. But, real data are messy and most of the time is spent cleaning up the data to make them usable. What are the problems here?\n",
    "\n",
    "- Sequences are not in a single line\n",
    "- There are blank spaces and new lines at end of each line\n",
    "\n",
    "How do we clean up the data and read it into the variables we have? We will use two tricks:\n",
    "\n",
    "- String manipulation for cleanup\n",
    "- Make loops and conditionals to read the file like a human"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### strip, rstrip and lstrip\n",
    "\n",
    "- Default - remove whitespace from beginning, end, or both sides \n",
    "- For rstrip, lstrip - all combinations of characters in the chars argument are removed from the right/left of the string until first mismatch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\tHuR\n",
      "Seq:\t\tST##MSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHS##EN\n",
      "Seq name:\tHuR\n",
      "Seq:\t\tST##LGYGFVNYVTAKDAERAINTLNGLRLQSKTIKVSYARPSSEVIKDANLYISGLPRTMTQK##EN\n",
      "Seq name:\tHuR\n",
      "Seq:\t\tST##DVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPIT##EN\n",
      "Seq name:\tHuR\n",
      "Seq:\t\tST##VKFAANPNQNKNVALLSQLYHSPARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNA##EN\n",
      "Seq name:\tHuR\n",
      "Seq:\t\tST##SSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAM##EN\n",
      "Seq name:\tHuR\n",
      "Seq:\t\tST##AIASLNGYRLGDKILQVSFKTNKSHK##EN\n",
      "Seq name:\tHistone_H3\n",
      "Seq:\t\tST##MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTE##EN\n",
      "Seq name:\tHistone_H3\n",
      "Seq:\t\tST##LLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTI##EN\n",
      "Seq name:\tHistone_H3\n",
      "Seq:\t\tST##MPKDIQLARRIRGERA##EN\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tST##MGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTV##EN\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tST##KAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFTTMEKAGAHLQGGAKRVI##EN\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tST##ISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHA##EN\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tST##ITATQKTVDGPSGKLWRDGRGALQNIIPASTGAAKAVGKVIPELNGKLTGMAFRVPTANV##EN\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tST##SVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAG##EN\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tST##IALNDHFVKLISWYDNEFGYSNRVVDLMAHMASKE##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##MCEEETTALVCDNGSGLCKAGFAGDDAPRAVFPSIVGRPRHQGVMVGMGQKDSYVGDEAQ##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##SKRGILTLKYPIEHGIITNWDDMEKIWHHSFYNELRVAPEEHPTLLTEAPLNPKANREKM##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##TQIMFETFNVPAMYVAIQAVLSLYASGRTTGIVLDSGDGVTHNVPIYEGYALPHAIMRLD##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##LAGRDLTDYLMKILTERGYSFVTTAEREIVRDIKEKLCYVALDFENEMATAASSSSLEKS##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##YELPDGQVITIGNERFRCPETLFQPSFIGMESAGIHETTYNSIMKCDIDIRKDLYANNVL##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##SGGTTMYPGIADRMQKEITALAPSTMKIKIIAPPERKYSVWIGGSILASLSTFQQMWISK##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##PEYDEAGPSIVHRKCF##EN\n",
      "Seq name:\tActin\n",
      "Seq:\t\tST##â€¨##EN\n"
     ]
    }
   ],
   "source": [
    "a_file = open('protein_sequences.fasta', 'r' )\n",
    "for i in a_file: #for each line\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        #Adding rstrip and lstrip\n",
    "        seq_name = i.rstrip().lstrip('>')\n",
    "    else:\n",
    "        #Adding rstrip\n",
    "        seq = i.rstrip()\n",
    "        #print the results inside the loop\n",
    "        print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\tST##\"+seq+\"##EN\")\n",
    "a_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, now the pesky newlines and spaces have been removed. What about the fact that sequences are on multiple lines? Here is where we make the code read the file like a human. If we were reading the file manually, here is how we would do it:\n",
    "\n",
    "- Is there a \">\" sign? That is a sequence name. The next line(s) will have the sequence.\n",
    "- No \">\" sign, so the previous sequence name still applies, let us keep appending the lines till we see a \">\" sign\n",
    "- We saw a \">\" sign again! The previous sequence is complete! Let us do the analysis on that previous sequence\n",
    "\n",
    "Here is how we an write the code for it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq name:\tHuR\n",
      "Seq:\t\tMSNGYEDHMAEDCRGDIGRTNLIVNYLPQNMTQDELRSLFSSIGEVESAKLIRDKVAGHSLGYGFVNYVTAKDAERAINTLNGLRLQSKTIKVSYARPSSEVIKDANLYISGLPRTMTQKDVEDMFSRFGRIINSRVLVDQTTGLSRGVAFIRFDKRSEAEEAITSFNGHKPPGSSEPITVKFAANPNQNKNVALLSQLYHSPARRFGGPVHHQAQRFRFSPMGVDHMSGLSGVNVPGNASSGWCIFIYNLGQDADEGILWQMFGPFGAVTNVKVIRDFNTNKCKGFGFVTMTNYEEAAMAIASLNGYRLGDKILQVSFKTNKSHK\n",
      "% of Positive Residues:14.1\n",
      "Seq name:\tHistone_H3\n",
      "Seq:\t\tMARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA\n",
      "% of Positive Residues:24.3\n",
      "Seq name:\tGapdh\n",
      "Seq:\t\tMGKVKVGVNGFGRIGRLVTRAAFNSGKVDIVAINDPFIDLNYMVYMFQYDSTHGKFHGTVKAENGKLVINGNPITIFQERDPSKIKWGDAGAEYVVESTGVFTTMEKAGAHLQGGAKRVIISAPSADAPMFVMGVNHEKYDNSLKIISNASCTTNCLAPLAKVIHDNFGIVEGLMTTVHAITATQKTVDGPSGKLWRDGRGALQNIIPASTGAAKAVGKVIPELNGKLTGMAFRVPTANVSVVDLTCRLEKPAKYDDIKKVVKQASEGPLKGILGYTEHQVVSSDFNSDTHSSTFDAGAGIALNDHFVKLISWYDNEFGYSNRVVDLMAHMASKE\n",
      "% of Positive Residues:13.7\n"
     ]
    }
   ],
   "source": [
    "ln=0 #We need to keep track of the line number\n",
    "seq=''\n",
    "a_file = open('protein_sequences.fasta', 'r' )\n",
    "for i in a_file: #for each line\n",
    "    ln+=1\n",
    "    if i.startswith('>'): #Does the line start with the \">\" symbol?\n",
    "        #Check if this is the first line:\n",
    "        if ln == 1:\n",
    "            pass\n",
    "        else:\n",
    "            #The previous sequence is complete, count the positive residues when we get to a sequence line\n",
    "            count = seq.count('K') + seq.count('R') + seq.count('H')\n",
    "            percent_positive = count*100/len(seq)\n",
    "            #print the results inside the loop\n",
    "            print(\"Seq name:\\t\"+seq_name+\"\\nSeq:\\t\\t\"+seq+\"\\n% of Positive Residues:\"+str(round(percent_positive,1)))\n",
    "            seq=''\n",
    "        \n",
    "        #Now process the new sequence name:\n",
    "        seq_name = i.rstrip().lstrip('>')\n",
    "        \n",
    "    else:\n",
    "        #print(seq_name,seq)\n",
    "        seq = seq + i.rstrip()\n",
    "\n",
    "a_file.close()\n",
    "#Does everything look good?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A real life problem\n",
    "\n",
    "Sequencing reads from Illumina platform are stored as fastq files. Each sequencing \"read\" gets four lines. Here is an example:\n",
    "\n",
    "`@SRR036139.1 11_26_8:5:1:749:1421 length=36\n",
    "GCTGGCTGCGTCTGTGGTGGGTTTCATGTTAAGGTC\n",
    "+SRR036139.1 11_26_8:5:1:749:1421 length=36\n",
    "IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII`\n",
    "\n",
    "Description from FASTQ wiki page:\n",
    " - Line 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line).\n",
    " - Line 2 is the raw sequence letters.\n",
    " - Line 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again.\n",
    " - Line 4 encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence.\n",
    " \n",
    "### We want to determine the GC content of the reads in any FASTQ file. \n",
    "GC content could tell you about any bias in the library or if you are sequencing a new species for example, it tells you the properties of the species' genome. So, if you have a fastq file, how would you determine the GC content of the reads?\n",
    "\n",
    "- Somehow get the sequences from the file into a data structure you learned\n",
    "- Examine the sequence to get GC content\n",
    "- Present the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To get GC content of the sequences in a fastq file:\n",
    " - Find the line with sequence : Count every 4th line starting from 2\n",
    " - Get the GC content : (number of G + number of C)/(total number of nucleotides) \n",
    "\n",
    "#### Before we begin, a couple of important points:\n",
    "\n",
    "#### Toy Models are Important\n",
    "When you write new code, don't apply it to actual data in the beginning - the data may be complicated enough for you to not be able to judge if your code is running properly. Instead, start with a toy model - something that will give you a trivial result. For example, here is a toy fastq file that has only two sequences with GC content of 100% and 50%:<br>\n",
    "`@TEST1\n",
    "GCGCGCGCGCGC\n",
    "+TEST1\n",
    "IIIIIIIIII\n",
    "@TEST2\n",
    "ATGCATGCATGC\n",
    "+TEST2\n",
    "IIIIIIIIIIII`\n",
    "\n",
    "#### Also important to spell out the architecture of the data structures you will create\n",
    "\n",
    "Some of the data structures we will use:\n",
    "\n",
    "- `seqs` - a dictionary, where the key will be the sequence number, value will be the sequence\n",
    "- `gc_content` - a dictionary, where the key will be the sequence number, value will be the GC content\n",
    "- `fastq_file` - a file wrapper for reading the fastq file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 3\n",
      "2 0\n",
      "3 1\n",
      "4 2\n",
      "5 3\n",
      "6 0\n",
      "7 1\n",
      "8 2\n",
      "9 3\n",
      "10 0\n",
      "11 1\n",
      "12 2\n",
      "13 3\n",
      "14 0\n",
      "15 1\n",
      "16 2\n",
      "17 3\n",
      "18 0\n",
      "19 1\n"
     ]
    }
   ],
   "source": [
    "#How do we get check for 2, 6, 10, 14th lines??\n",
    "#If we subtract 2 from them: 0, 4, 8, 12, ... should be divisible by 4. Let us check:\n",
    "for i in range(1,20):\n",
    "    print(i,(i-2)%4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 is the sequence line\n",
      "6 is the sequence line\n",
      "10 is the sequence line\n",
      "14 is the sequence line\n",
      "18 is the sequence line\n"
     ]
    }
   ],
   "source": [
    "#Can we now pick the line numbers with the sequence?\n",
    "for i in range(1,20,1):\n",
    "    if ((i-2)%4 == 0):\n",
    "        print(i,\"is the sequence line\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "line 2 contains the sequence for read # 1\n",
      "line 6 contains the sequence for read # 2\n",
      "line 10 contains the sequence for read # 3\n",
      "line 14 contains the sequence for read # 4\n",
      "line 18 contains the sequence for read # 5\n"
     ]
    }
   ],
   "source": [
    "#Ok, line 2 is the sequence line for the 1st sequence, 6 for second sequence, and so forth.\n",
    "#Can we assign this explicitly?\n",
    "for i in range(2,20,4):\n",
    "    key = int((i-2)/4)+1 # Convert the line_no to a sequence ID\n",
    "    print(\"line\",i,\"contains the sequence for read #\",key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, let us put all of this together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 GCGCGCGCGCGC 100.0\n",
      "3 ATGCATGCATGC 50.0\n"
     ]
    }
   ],
   "source": [
    "seqs={} # A dictionary to store sequences if needed later\n",
    "\n",
    "gc_content={} # A dictionary to store GC content if needed later\n",
    "\n",
    "line_no = 0 # Variable to count lines\n",
    "\n",
    "fastq_file = open('toy.fastq', 'r' ) #Open the fastq file\n",
    "\n",
    "for line in fastq_file: # Iterate through the fastq file\n",
    "    line_no+=1 #counting line numbers\n",
    "      \n",
    "    if (line_no-2)%4 == 0: # Does the line number belong to the series 2, 6, 10, 14, etc. ?\n",
    "        key = int((line_no-2)/4)+1 # Convert the line_no to a sequence ID\n",
    "        line=line.rstrip()\n",
    "        seqs[key] = line #Populate seqs dictionary\n",
    "        gc = line.count('G') + line.count('C')\n",
    "        gc=round(gc*100/len(line),2) #make it percentage, make it look pretty with round()\n",
    "        gc_content[key] = gc\n",
    "        print(key+1,line,gc) #print the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us try the real data now. (First 250 sequences from the SRA file: SRR036139"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs={} # A dictionary to store sequences if needed later\n",
    "\n",
    "gc_content={} # A dictionary to store GC content if needed later\n",
    "\n",
    "line_no = 0 # Variable to count lines\n",
    "\n",
    "fastq_file = open('example.fastq', 'r' ) #Open the fastq file\n",
    "\n",
    "#Let us write sequence details in a file called \"seq_details.txt\"\n",
    "fh = open(\"seq_details.txt\",\"w\")\n",
    "\n",
    "for line in fastq_file: # Iterate through the fastq file\n",
    "    line_no+=1 #counting line numbers\n",
    "      \n",
    "    if (line_no-2)%4 == 0: # Does the line number belong to the series 2, 6, 10, 14, etc. ?\n",
    "        key = int((line_no-2)/4)+1 # Convert the line_no to a sequence ID\n",
    "        line=line.rstrip()\n",
    "        seqs[key] = line #Populate seqs dictionary\n",
    "        gc = line.count('G') + line.count('C')\n",
    "        gc_p=round(gc*100/len(line),2) #make it percentage, make it look pretty with round()\n",
    "        gc_content[key] = gc_p\n",
    "        #print the results to a string\n",
    "        write_str = \"Seq #\"+ str(key) + \" Seq: \" + line + \" Number of GC: \" + str(gc) +\" Sequence Length: \"+ str(len(line)) \\\n",
    "        + \" GC Content: \" + str(gc_p) + \"\\n\"\n",
    "        #print the string to a file\n",
    "        fh.write(write_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the \"seq_details.txt\" file - we just calculated the GC content of 250 sequences.\n",
    "\n",
    "However, this is not really useful yet because there are 250 numbers and we don't have a sense of what the data is telling us. A histogram may be more useful. There is a direct histogram function we will discuss later, but here will do it from scratch. This is an example of how powerful coding can be - writing code to (insert your favorite analysis) from scratch!\n",
    "\n",
    "### How to make a histogram? \n",
    "\n",
    "Wikipedia says: \"To construct a histogram, the first step is to \"bin\" (or \"bucket\") the range of values—that is, divide the entire range of values into a series of intervals—and then count how many values fall into each interval. The bins are usually specified as consecutive, non-overlapping intervals of a variable. The bins (intervals) must be adjacent, and are often (but not required to be) of equal size.\"\n",
    "\n",
    "#### How to bin data?  \n",
    "\n",
    "The idea is to put an interval of numbers into the midpoint like this:\n",
    "\n",
    "1 2 3 4 5 6 7 8 9 10,11,12...\n",
    "\n",
    "If we group these numbers by 4:\n",
    "\n",
    "(1,2,3,4) (5,6,7,8) (9,10,11,12)\n",
    "\n",
    "The midpoints are:\n",
    "\n",
    "2.5, 6.5, 10.5\n",
    "\n",
    "If we group the same numbers by 5:\n",
    "\n",
    "(1,2,3,4,5) (6,7,8,9,10) (11,12,13,14,15)\n",
    "\n",
    "The midpoints are:\n",
    "3, 8, 13\n",
    "\n",
    "#### How do we generalize this in our code? \n",
    "By combining `int()` and `round()`. I came up with the formula below by trial and error. I am sure you can come up with your own and even do better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "i bin\n",
      "1 3.0\n",
      "2 3.0\n",
      "3 3.0\n",
      "4 3.0\n",
      "5 3.0\n",
      "6 8.0\n",
      "7 8.0\n",
      "8 8.0\n",
      "9 8.0\n",
      "10 8.0\n",
      "11 13.0\n",
      "12 13.0\n",
      "13 13.0\n",
      "14 13.0\n",
      "15 13.0\n",
      "16 18.0\n",
      "17 18.0\n",
      "18 18.0\n",
      "19 18.0\n",
      "20 18.0\n"
     ]
    }
   ],
   "source": [
    "bin_size = 5\n",
    "fact = (bin_size/2)-0.5\n",
    "print(\"i bin\")\n",
    "for i in range(1,21,1):\n",
    "    #bin1 = int( round( ((i-0.5)/bin_size + 0.5),0))*bin_size\n",
    "    #bin1 = int( round((i/bin_size),0))*bin_size\n",
    "    bin1 = int( round(( (i-0.5)/bin_size + 0.5),0))*bin_size - fact\n",
    "    print(i,bin1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, we can assign a \"bin\" to any number. Now we need to count the numbers in each bin. We will put these counts in a dictionary called `hist`, where the keys are the bins and values are the counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.0 5\n",
      "8.0 5\n",
      "13.0 5\n",
      "18.0 5\n"
     ]
    }
   ],
   "source": [
    "hist = {}\n",
    "\n",
    "bin_size = 5\n",
    "fact = (bin_size/2)-0.5\n",
    "\n",
    "for i in range(1,21,1):\n",
    "    bin1 = int( round(( (i-0.5)/bin_size + 0.5),0))*bin_size - fact\n",
    "    #if the bin exists, add to it\n",
    "    if bin1 in hist:\n",
    "        hist[bin1] += 1\n",
    "    #if bin doesn't exist, add to it\n",
    "    else:\n",
    "        hist[bin1] = 1\n",
    "\n",
    "#let us now print the histogram\n",
    "for i in sorted(hist):\n",
    "    print(i,hist[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, let us do the same with our `gc_content` dictionary. The keys of `gc_content` are sequence numbers and the values are the corresponding GC content in percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#let us assume bin size of 5 percent\n",
    "bin_size = 2\n",
    "fact = (bin_size/2)-0.5\n",
    "\n",
    "#let us write the histogram into a file called \"GC_hist.txt\"\n",
    "fh = open(\"GC_hist.txt\",\"w\")\n",
    "\n",
    "hist = {} #declare the histogram\n",
    "\n",
    "for i in gc_content:\n",
    "    bin1 = int( round(( (gc_content[i]-0.5)/bin_size + 0.5),0))*bin_size - fact\n",
    "    if bin1 in hist:\n",
    "        hist[bin1] += 1\n",
    "    else:\n",
    "        hist[bin1] = 1\n",
    "for i in sorted(hist):\n",
    "    hist[i]=hist[i]*100/len(gc_content)\n",
    "    write_str =str(i) + \" \" + str(hist[i]) + \"\\n\"\n",
    "    fh.write(write_str)\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would be nice if we could see a plot. I don't know how to make a simple plot in Jupyter, so I just googled and found this link: https://matplotlib.org/gallery/lines_bars_and_markers/simple_plot.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEGCAYAAAB/+QKOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3dd3zU9f3A8dc7e5BBSCBhz4CIgBhxAoqooNbdFqqtq2Kts3ba/rTLto621tFaqYq27lYcrSBQpSAOJCDIHiYBAmQCCSRk3vv3x10whku4JHf3vUvez8cjD+477r5vvpC88/2st6gqxhhjTEsRTgdgjDEmNFmCMMYY45UlCGOMMV5ZgjDGGOOVJQhjjDFeRTkdgD+lp6fr4MGDnQ7DGGPCxqpVq8pUNcPbsS6VIAYPHkxubq7TYRhjTNgQkR2tHbMmJmOMMV5ZgjDGGOOVJQhjjDFeWYIwxhjjlSUIY4wxXlmCMMYY45UlCGOMMV5ZgjDGdNj+qjpeXbmLmvpGp0MxAWAJwhjTYS+s2MGPXvuMGY+8z4q8cqfDMX5mCcIY02F5pVUkx0XR4HLx9Tkfc88b6zlYU+90WMZPLEEYYzosv7yKMf1SWHjnZG44cwjPr9jB+Q8vY8mWEqdDM34QsAQhIs+ISImIrG+27xURWeP5KhCRNa28t0BE1nnOs8WVjAlR+WVVDE5PJCEminsuGs1rN59OYmwU181dyV2vrGF/VZ3TIZpOCOQTxLPA9OY7VPXrqjpeVccDrwHz2nj/2Z5zcwIYozGmg/ZX1XGgup6h6YlH9k0Y2JP/3H4mt58zgrfW7mHaH5fyn8/2oKoORmo6KmAJQlWXAfu8HRMRAb4GvBSo6xtjAiu/vAqAwb0Sv7Q/NiqSu87N5t+3nUm/nvHc+uKn3PSPVRRX1jgRpukEp/ogJgHFqrqtleMKLBKRVSIyu60PEpHZIpIrIrmlpaV+D9QY411+qTtBDMlI9Hr8uKxk5t18Oj+9YBRLt5Yy7Y9LeWXlTnuaCCNOJYhZtP30cIaqTgBmALeIyOTWTlTVOaqao6o5GRlea14YYwKgoLyKCIEBPRNaPScqMoLZk4ex8M7JjM5K5sevreOqp1aws7w6iJGajgp6ghCRKOBy4JXWzlHVPZ4/S4DXgYnBic4Y46u8sioGpCUQE3XsHyOD0xN56cZT+c1lY/issILz/7SMp5fn0+iyp4lQ5sQTxDRgs6oWejsoIokiktT0GjgPWO/tXGOMcwrKqo7qf2hLRIRw1SmDWHzXZE4b1otf/2cjVzzxIVuLDwYwStMZgRzm+hLwETBSRApF5AbPoZm0aF4Skb4iMt+z2QdYLiJrgU+At1X1nUDFaYxpP1Ulv6yKIem+J4gmWSnxPH1NDo/MHM/OfdVc+Oj7PPruNuoaXAGI1HRGwGpSq+qsVvZf62XfHuACz+s8YFyg4jLGdF7pwVqq6xo7lCAARIRLxvfjzOHp/PLfG/nj4q3MX7eXB64Yy7gBqX6O1nSUzaQ2xrRbXplnBFMHE0STXj1ieXTWiTz1rRwOVNdz2V8+4HfzN3G4zhb/CwWWIIwx7VbgpwTRZNroPiy6azJfP3kgTy7LY8Yjy/jYFv9znCUIY0y75ZdVERMZQd/UeL99ZnJcNL+7/ARevPEUFJg552N+9vo6W/zPQZYgjDHtll9WxcBeCURGiN8/+/Rh6bxzx2RunDSElz7ZyXkPL+O9zcV+v445NksQxph26+gIJl/Fx0TyswtHM++7Z5AcF831z+Zyx8ufUn6oNmDXNEezBGGMaZdGl7JjX/WXFukLlPEDUvn3bWfyvWnZzF+3l3MfXsbybWUBv65xswRhjGmXPQcOU9fgYnAQEgRATFQEd0wbwdu3TyIhJpI//XdrUK5rLEEYY9op388jmHyV3SeJiYPT2Fthq8IGiyUIY0y7FJQ7kyAAslLjKK6ssTWcgsQShDGmXfJKq0iIiaR3UmzQr52ZEk+DS62zOkgsQRhj2qWg3L1In7vuV3BlJccBsMeamYLCEoQxpl3yy6paLRIUaJkp7gRRVHHYket3N5YgjDE+q2twUbj/MEPascy3P2V5EoR1VAeHJQhjjM927a+m0aWOdFADpCXGEBMVQZEliKCwBGGM8VnTIn3BmgPRkoiQlRJnTxBBYgnCGOOzpjkQwZhF3ZrM5Dh7gggSSxDGGJ/ll1WREh9Nz8QYx2LISoljb6V1UgeDJQhjjM8CvUifLzJT4imqqMFlk+UCzhKEMcZnBWVVjjYvgfsJor5RKa+qczSO7iBgCUJEnhGREhFZ32zfL0Rkt4is8Xxd0Mp7p4vIFhHZLiI/CVSMxhjfHa5rZE9FjWMd1E2+mAth/RCBFsgniGeB6V72P6yq4z1f81seFJFI4M/ADGA0MEtERgcwTmOMD3bsc24Npua+mAth/RCBFrAEoarLgH0deOtEYLuq5qlqHfAycIlfgzPGtFt+aagkCHeZ06JKe4IINCf6IG4Vkc88TVA9vRzvB+xqtl3o2eeViMwWkVwRyS0tLfV3rMYYjzyH50A06ZUYQ3Sk2FyIIAh2gngCGAaMB/YCf/ByjrcVwFodrqCqc1Q1R1VzMjIy/BOlMeYoBWVVZCTF0iM2ytE4IiKEPslx7D1gTUyBFtQEoarFqtqoqi7gb7ibk1oqBAY02+4P7AlGfMaY1oXCENcmNps6OIKaIEQkq9nmZcB6L6etBEaIyBARiQFmAm8FIz5jTOsKyqscW6SvpcyUeOuDCIJADnN9CfgIGCkihSJyA/CgiKwTkc+As4Hvec7tKyLzAVS1AbgVWAhsAl5V1Q2BitMYc2yVNfWUHapzbJnvlpqeIFRtslwgBawxUVVnedn9dCvn7gEuaLY9HzhqCKwxxhlHFukLkSeIrJQ46hpc7K+uJ83BZT+6OptJbYw5piOL9IXQEwTYXIhAswRhjDmm/LIqRGBgWoLToQDuPgiw2dSBZgnCGHNM+WVV9E2JJy460ulQgC+eIKw2dWBZgjDGHFNBWVXINC8BpPeIJTJCrDZ1gFmCMMa0SVXJK6sKmQ5qgMgIoU9SrM2FCDBLEMaYNu2rquNgTUPITJJrkplileUCzRKEMaZNTSOYQi1BZKXGW4IIMEsQxpg2hWyCSLbJcoFmCcIY06b8siqiIoT+PeOdDuVLMlPiOFzfSMXheqdD6bIsQRhj2pRfVsXAtASiIkPrx0VTXQjrqA6c0PoXN8aEnPyyKsdrQHhjpUcDzxKEMaZVLpe6V3ENwQTxxXIbliACxRKEMaZVxQdrqKl3heQTRO+kWCIEmywXQJYgjDGtaqpDPTQEE0RUZAS9k6xwUCBZgjDGtCq/PDTqULcm0yrLBZQlCGNMq/JLq4iNiiArOc7pULxyFw6yJqZAsQRhjGlVUwd1RIQ4HYpXmVZZLqAsQRhjWhVqi/S1lJUSR3VdIwdrG5wOpUuyBGGM8aqh0cWufdUhU4famywrHBRQAUsQIvKMiJSIyPpm+x4Skc0i8pmIvC4iqa28t0BE1onIGhHJDVSMxpjW7T5wmPpGZUiIP0GAzYUIlEA+QTwLTG+xbzEwRlXHAluBu9t4/9mqOl5VcwIUnzGmDUcW6QvhJ4gvZlNbR3UgtCtBiEhPERnry7mqugzY12LfIlVtaiz8GOjfnusbY4KnKUGEch9E76Q4RGDPAXuCCIRjJggR+Z+IJItIGrAWmCsif/TDta8HFrRyTIFFIrJKRGYfI77ZIpIrIrmlpaV+CMsYA+4yo0mxUaT3iHE6lFbFREWQ3iPW+iACxJcniBRVrQQuB+aq6knAtM5cVER+BjQAL7RyyhmqOgGYAdwiIpNb+yxVnaOqOaqak5GR0ZmwjDHN5HkW6RMJzSGuTbJS4thbaQkiEHxJEFEikgV8DfhPZy8oItcAFwFXaSuDl1V1j+fPEuB1YGJnr2uMaZ/8stBcpK+lzOQ464MIEF8SxK+AhcDnqrpSRIYC2zpyMRGZDvwYuFhVq1s5J1FEkppeA+cB672da4wJjNqGRnYfOByyS2w01zc13kYxBcgxE4Sq/lNVx6rqzZ7tPFW94ljvE5GXgI+AkSJSKCI3AI8DScBizxDWv3rO7Ssi8z1v7QMsF5G1wCfA26r6Tof+dsaYDtlZXo1qaC7S11JmShwHaxo4ZJPl/C7qWCeISDbwBNBHVcd4RjFdrKr3tfU+VZ3lZffTrZy7B7jA8zoPGHesuIwxgXNkBFMYJIisZkNdh/dOcjiarsWXJqa/4Z6vUA+gqp8BMwMZlDHGWUfmQITwENcmmck2WS5QfEkQCar6SYt99ixnTBdWUF5Fr8QYUhKinQ7lmKw2deD4kiDKRGQY7rkJiMiVwN6ARmWMcVReaWjWofamT0osYOsxBcIx+yCAW4A5wCgR2Q3kA1cHNCpjjKMKyquYNCI85hXFRkWS3iPGniAC4JgJwtNpPM0z5DRCVQ8GPixjjFOqahsorqwNizkQTTJTbC5EIPiy1MZvRSRVVatU9aBnPaY2RzAZY8JXgafMaFgliGSbCxEIvvRBzFDVA00bqrofz5BUY0zXEw6L9LWUZbWpA8KXBBEpIrFNGyISD8S2cb4xJowVHJkDkeBwJL7LTImj4nA91XU2wNKffOmkfh54V0Tm4h7JdD3wXECjMsY4Jq+siszkOBJifPnxEBq+mCxXw9CMHg5H03X40kn9oIisA84BBPi1qi4MeGTGGEeEyyJ9zTUvPWoJwn98+hVBVRfQeu0GY0wXUlBWxfQxWU6H0S5WejQwfBnFdLmIbBORChGpFJGDIlIZjOCMMcF1oLqO/dX1YbFIX3NHSo9aXQi/8uUJ4kHgK6q6KdDBGGPcVJU5y/JIiY/m8gn9iYkKZPn4L4TTIn3NxUVH0jMhmj0HbC6EP/mSIIotORgTXIs2FvO7BZsBePTdbXznrGF8LWcAcdGRAb3ukUX6wixBAGSmxNtyG37my68luSLyiojM8jQ3XS4ilwc8MmO6qaraBn7x1gZGZSYx99qTyUqN5943NzDpwSU89X5eQIdyFpRVESEwMC18hrg2sbkQ/ufLE0QyUI27slsTBeYFJCJjurk//XcreytqePwbJ3LSoDTOGpnBR3nlPPbudu57exNP/O9zbpg0hG+eOoikOP+utppXVkX/nglBa9Lyp8yUONbsOnDsE43PfBnmel0wAjHGwKa9lTzzQQGzJg7gpEFpAIgIpw9L5/Rh6eQW7OOx97bz4DtbeHJpHtedMZjrTh/it2W5C8rDb4hrk74pceyrqqOmvjHgTXHdhS+jmLJF5F0RWe/ZHisi/xf40IzpXlwu5WevryMlPpofTx/l9ZycwWk8d/1E3rzlDCYOSeNP/93GmQ+8x0MLN7Ovqq5T11dV8kvDN0FkeuZCFNtIJr+xinLGhIhXcnexeucBfnrBcaQmxLR57rgBqfztWznMv30Sk7Mz+Mv/PueM+9/jN29vpORgx35Alh6qpaquMWwThM2F8L+AVpQTkWdEpKTp6cOzL01EFnvmViwWkZ6tvPcazznbROQaX65nTLgqP1TL/Qs2c8qQNK6Y0M/n943um8yfr5rAojsnM31MJk8vz2fSA0v4xVsb2j3kM780PIe4Nsk8kiBsqKu/BLqi3LPA9Bb7fgK8q6ojgHc9218iImnAz4FTgInAz1tLJMZ0Bb+dv5mq2gbuu3QMItLu94/ok8TDXx/Pe98/i0vG9+X5j3cw5aEl3D1vHbv2Vfv0GU3LfIfbJLkmVpva/3xJELcAT/JFRbk7gZt9+XBVXQbsa7H7Er5Y7O854FIvbz0fWKyq+zzLiy/m6ERjTJfwcV45r60uZPbkoYzok9SpzxqcnsiDV45jyQ/O4usnD+C1VYWc9fv/8YN/riWv9FCb780rqyImMoK+qfGdisEpibFRJMdF2VwIP3KiolwfVd3r+ey9ItLbyzn9gF3Ntgs9+44iIrOB2QADBw7sZGjGBFddg4v/e2M9/XvGc9vUEX773AFpCdx36QncevYI5izL48VPdjBvdSEXje3LrVOHk+0lERWUVTGwVwKREe1/ggkVfVOtcJA/HTNBiMi9LbYBUNVfBSgmcK8a25J6O1FV5+CumU1OTo7Xc4wJVX97P4/tJYeYe+3JxMf4f2hmZkoc935lNDefNYynlufxj4928NbaPUw/PpNbpw5nTL+UI+fml1WFVZEgb9ylRy1B+IsvTUxVzb4agRnA4E5cs1hEsgA8f5Z4OacQGNBsuz+wpxPXNCbk7NpXzWPvbWP68ZmcPcrbg7T/ZCTFcveM4/jgx1O5fepwPvi8jIseW84Nz67k0537cbmUgvJqhmaEd4Kw2dT+5UsT0x+ab4vI74G3OnHNt4BrgPs9f77p5ZyFwG+bdUyfh3uorTFdgqpy75vriRTh5xePDtp1eybGcNd5I7lh0lD+/mEBT3+Qz2V/+ZCcQT2pa3CF/xNEcjxlh2qpbWgkNsomy3VWR+bTJwBDfTlRRF4CPgJGikihiNyAOzGcKyLbgHM924hIjog8BaCq+4BfAys9X7/y7DOmS1i4oYglW0r53rnZR4rdBFNKfDS3nTOCD348lbtnjDoygmlkZngX22maC1FSWetwJF2DL30Q6/ii/T8SyAB86n9Q1VmtHDrHy7m5wLebbT8DPOPLdYwJJ4dqG/jlvzdyXFYy154+2NFYEmOjuGnKML512mA27KlgwsDwHk2e2Wyy3IAwXHAw1PiyWN9FzV434F7+2yqDG9NBf1q8laLKGv581QSiIkNjUbz4mEhyBqc5HUan9U21yXL+5EuCaDmsNbn5RB5r+jHGdxv3VDL3wwJmTRwY9r+th6LMZrWpTef5kiBW4x5RtB/38NNUYKfnmOJjf4Qx3Z3LpfzsjXWkxkfz4/O9L8ZnOqdHbBRJsVE2kslPfHm+fQd3ydF0Ve2Fu8lpnqoOUVVLDsb46OWVu/h05wF+duFxflue2xwtMyXOmpj8xJcEcbKqzm/aUNUFwJTAhWRM11N2qJb7F2zi1KFpXHai74vxmfazyXL+4+tiff8nIoNFZJCI/AwoD3RgxnQlv52/icP1jdx36QkdWozP+M4my/mPLwliFu6hra97vjI8+4wxPvjo83Lmrd7NTZOHMbx3eM8zCAeZKfGUHqqlvtHldChhz5eZ1PuAO0Skh6q2vRykMeZL8suquOvVNQxIi+fWqcOdDqdb6JsShyqUHKylX5iuTBsqfCk5erqIbAQ2erbHichfAh6ZMWFue8khvv7kR9Q2uHjy6hyrkxwkTZPliqyjutN8aWJ6GHd9hnIAVV0LTA5kUMaEuy1FB5k55yNcCi/PPpXRfZOdDqnbaFq6xPohOs+naZyquqvFrsYAxGJMl7BhTwUz53xEZITwyk2neq29YALnyHIbByxBdJYvE+V2icjpgIpIDHA7sCmwYRkTnj4rPMA3n/6ExJhIXrzx1LCt7xzOkuOiSIiJtCcIP/DlCeI7uMuO9sNdp2G8Z9sY08yqHfu56m8rSI6P4pWbTrPk4BARcc+FqLQ+iM7yZRRTGXBVEGIxJmytyCvn+mdX0js5jhe+fUrY1nXuKvqmWOlRf/BlFNODIpIsItEi8q6IlInI1cEIzphw8MH2Mq6du5LMlDhemX2qJYcQYLOp/cOXJqbzVLUS9xpMhUA28MOARmVMmFi6tZTrn13JwLQEXp59Gr2T45wOyeCeTV1ysJYGmyzXKb4kiKZVxS4AXrLlvY1xe3dTMTc+l8uwjB68NPtUMpJinQ7JeGSmxNHoUkoPWWW5zvAlQfxbRDYDOcC7IpIB2LOb6dbeWV/Ed55fxaisJF688RTSEmOcDsk0k9WsspzpuGMmCFX9CXAakKOq9UA1cEmgAzMmVP177R5ueXE1J/RL4flvn0JqgiWHUJOZbIWD/MHXiXL7VbXR87pKVYs6ekERGSkia5p9VYrInS3OOUtEKpqdc29Hr2eMP81bXcgdL3/KSYN68vcbTiE5zuo6hKIvSo9agugMXybK+ZWqbsE9lwIRiQR2414ltqX3VfUiL/uNCSpVZW1hBf9eu4dnPsjntKG9eOqaHBJigv7tY3yUEh9NXHSErcfUSa3+DxeRM1T1AxGJVdVA9fScA3yuqjsC9PnGdIjLpazauZ/56/aycH0ReypqiIoQLh7XlweuGGsL74U4ESHL5kJ0Wlu/Aj0KnAR8BEwI0PVnAi+1cuw0EVkL7AF+oKobvJ0kIrOB2QADBw4MSJCme2hodLEifx8L1u9l4YZiSg/WEhMVweQRGXz/vJFMO66PlQoNI5nJNheis9pKEPUiMhfoJyKPtjyoqrd35sKedZ0uBu72cng1MEhVD4nIBcAbwAhvn6Oqc4A5ADk5OdqZmEz3U9fg4oPtZSxYv5fFG4vZX11PfHQkZ4/KYPqYLKaO6k2PWGtKCkdZKXGsyLdR+Z3R1v/8i4BpwFRgVQCuPQNYrarFLQ94JuY1vZ4vIn8RkXTPsh/GdEpNfSNLt5ayYN1e3t1UwsHaBpJiozjnuN5MH5PFlOwM4mOsCSncZabEUVxZQ6NLiYywMq8d0WqC8PwwfllENnlqQPjbLFppXhKRTKBYVVVEJuIebWV1sE2HHaptYMnmEt5ZX8SSLSVU1zWSmhDN9DGZzDghkzOGpxMbZUmhK8lKiaPBpZQfqrUZ7h3ky7NzuYi8DpwBKLAcuENVCzt6URFJAM4Fbmq27zsAqvpX4ErgZhFpAA4DM1XVmo9Mu1RU1/PfTcUsWF/Esm2l1DW4SO8Ry2Un9mPGmCxOGZpGdKRPI71NGGpeOMgSRMf4kiDmAi8CX/VsX+3Zd25HL6qq1UCvFvv+2uz148DjHf18032VH6pl8UZ3UvhgexkNLiUrJY6rThnIjDFZnDSopzU3dBOZzWZTjxvgcDBhypcE0VtV5zbbfrblxDZjnFRcWcPCDUUsWFfEivxyXAoD0xK44cwhTB+Tybj+qURYUuh2sqw2daf5kiBKPct7N/UXzML6A0wI2LWvmu+/upaVO/ahCsN79+CWs4czfUwmo7OSEbGk0J2lJcYQExlhcyE6wZcEcT3u5p6HcfdBfOjZZ4xjauobuekfqyjcX833pmUzY0wmI6z2s2mmqbKcJYiO86Wi3E7c8xWMCRk/f3MDG/dW8sy1OUwd1cfpcEyIssJBnWNDOEzYeTV3F6/k7uKWs4dZcjBt6psSx16rTd1hliBMWNmwp4J73ljP6cN6cde5I50Ox4S4zJR4iitqcblslHxHWIIwYaPicD3ffWE1qQnRPDrrRBuuao4pKyWOukYX+6rrnA4lLPmcIETkVBF5T0Q+EJFLAxmUMS2pKj/851p27z/MX66aQHoPK+9pji3zyFBX64foiFYThGe5i+buwt1ZPR34dSCDMqalOcvyWLSxmLsvOI6TBqU5HY4JE01zIfYcsH6IjmhrFNNfRWQV8JCq1gAHgG8ALqCyjfcZ41cf55Xz4MItXHhCFtefMdjpcEwYOfIEUWlPEB3R6hOEql4KrAH+IyLfBO7EnRwSAGtiMkFRUlnDrS9+yqC0BO6/4gSb/GbaJT0xluhIsbkQHdRmH4Sq/hs4H0gF5gFbVPVRVS0NRnCme2todHHrS59SVdvAE1efRJLVfzbtFBEh9LHCQR3WVh/ExSKyHHgPWI+7+ttlIvKSiAwLVoCm+3po4RY+yd/Hby8fw8hMmyVtOiYrJY69th5Th7TVB3EfcBoQD8xX1YnAXSIyAvgN7oRhTEAs3FDEk8vyuOqUgVx2Yn+nwzFhLDMlnnWFB5wOIyy1lSAqcCeBeKCkaaeqbsOSgwmggrIqfvDqWsb2T+Her4x2OhwT5rJS4li0oQZVtT6sdmqrD+Iy3B3SDbhHLxkTcDX1jdz8wmoiIoQ/f2OCVXkznZaZHEdtg4v91fVOhxJ2jlVy9LEgxmIM97yxnk17K5l77ckMSEtwOhzTBWQdKRx0mLTEGIejCS+21IYJGa+s3Mk/VxVy29ThnD2qt9PhmC4iK9VdetRGMrWfJQgTEtbvruCeNzdw5vB07pyW7XQ4pgvJalZ61LSPLwWDAkJECoCDQCPQoKo5LY4L8AhwAVANXKuqq4Mdp/FdXYOLX/9nI9tKDrb7vdtLDpGWEMMjM8fbInzGr9J7xBIZITbUtQMcSxAeZ3v6OryZAYzwfJ0CPOH504QgVeUnr33GvE93kzOoZ7trQB+XlcwPzx9JL1uEz/hZZIQwoncP1uyyoa7t5XSCaMslwN9VVYGPRSRVRLJUda/TgZmj/WHRVuZ9upvvn5vNbeeMcDocY75kcnYGz35QQFVtA4mxofxjL7Q42QehwCIRWSUis70c7wfsarZd6Nn3JSIyW0RyRSS3tNRWAHHCiyt28viS7cw8eQC3Th3udDjGHGVKdgZ1jS4+zit3OpSw4mSCOENVJ+BuSrpFRCa3OO6tjeKoslCqOkdVc1Q1JyMjIxBxmjYs2VzCPW+u5+yRGdx36RibiGRCUs7gnsRHR7Jsq/0S2R6OJQhV3eP5swR4HZjY4pRCYECz7f7AnuBEZ3zxWeEBvvvCakZnJfP4NyYQFWmD4kxoio2K5LRhvVhqCaJdHPmOFpFEEUlqeg2ch3tBwObeAr4lbqcCFdb/EDp27avm+mdX0qtHDE9fm2PtuibkTcnOoKC8mh3lVU6HEjac+pWvD7BcRNYCnwBvq+o7IvIdEfmO55z5QB6wHfgb8F1nQjUt7a+q45q5n1DfqDx73UR6J8U5HZIxxzQ5290Ebc1MvnPk1z5VzQPGedn/12avFbglmHGZY6upb+TGv+dSuP8wL3z7FIb37uF0SMb4ZHCvBAamJbB0aynfPG2w0+GEBWs0Nj5zuZS7Xl3Dqp37efhr4zl5sNWGNuFDRJiSncGHn5dT1+ByOpywYAnC+Ow38zcxf10RP7vgOC4cm+V0OMa02+TsDKrrGsndsc/pUMKCJQjjk2eW5/P08nyuPX0wN5w5xOlwjOmQ04b1IjpSbDSTjyxBmGNasG4vv357I+cf34d7Lhptcx1M2OoRG0XOoDSWbrEE4QtLEKZNuQX7uPOVNZw4IJVHZp5oC+mZsDc5O4PNRQcprrTVXY/FEoRp1eelh/j233Ppmw9T74cAAA9gSURBVBrPU9ecTFy0VXcz4W+KDXf1mSUI41XpwVqunfsJkSI8e93JVonLdBnHZSWRkRRr/RA+sARhjlJd18ANz62k9GAtT197MoN6JTodkjF+IyJMHpHB8u1lNLqOWt7NNGMJwnxJQ6OL2178lPW7K3h81gTGD0h1OiRj/G7KyAwOVNfzWaHViGiLJQhzhKpy71sbeHdzCb+6ZAzTRvdxOiRjAmLS8HREsGamY7AEYY74y/8+58UVO7n5rGFcfeogp8MxJmB6JsYwtn+qdVQfgyUIA8Abn+7moYVbuGR8X3543kinwzEm4KZkZ7Bm1wEOVNc5HUrIsgRh+HB7GT/811pOHZrGg1eObXc9aWPC0ZTsdFwKy7eXOR1KyLIE0c1tLqrkpn+sYkh6Ik9+M4fYKJvrYLqHcf1TSY6LsmamNliC6Mb2VhzmurkrSYiN5NnrJpISH+10SMYETVRkBJNGZLB0aynu6gKmJUsQ3dTBmnqum7uSgzUNzL12In1T450OyZigm5ydTnFlLVuKDzodSkiyBNEN1TW4uPn51WwvOcQTV09gdN9kp0MyxhFWZa5tliC6GVXlJ/M+Y/n2Mu6/YiyTRmQ4HZIxjslKiWdknySbD9EKSxDdzB8Xb2Xe6t3cdW42V57U3+lwjHHc5Ox0Vubvp7quwelQQk7QE4SIDBCRJSKySUQ2iMgdXs45S0QqRGSN5+veYMfZFb30yU4ee287M08ewG1ThzsdjjEhYUp2b+oaXXycV+50KCEnyoFrNgDfV9XVIpIErBKRxaq6scV576vqRQ7E1yUt2VzC/72xninZGfz60jFW9McYj5zBPYmPjmTpllKmjrLlZZoL+hOEqu5V1dWe1weBTUC/YMfRnawrrOCWF1czKjOJP181gehIa1k0pklcdCSnDk2zfggvHP1JISKDgROBFV4OnyYia0VkgYgc38ZnzBaRXBHJLS21f+CWdpRXcd2zK+mZEMPca0+mR6wTD43GhLYp2RkUlFezo7zK6VBCimMJQkR6AK8Bd6pqZYvDq4FBqjoOeAx4o7XPUdU5qpqjqjkZGTYip0llTT0PLdzM+X9aRl1DI89dfzK9k+OcDsuYkDRlZG/Ahru25EiCEJFo3MnhBVWd1/K4qlaq6iHP6/lAtIikBznMsFTb0MjTy/OZ8uAS/rzkc84/PpO3b5/E8N5JTodmTMga3CuBAWnx1szUQtDbG8TdO/o0sElV/9jKOZlAsaqqiEzEnchsiEEbXC7lrbV7+P2iLRTuP8yZw9P5yYxRjOmX4nRoxoQ8EWFKdgbzVu+mrsFFTJT104Ezo5jOAL4JrBORNZ59PwUGAqjqX4ErgZtFpAE4DMxUWyylVcu2lnL/gs1s3FvJ8X2T+d3lJ9gEOGPaaUp2b57/eCe5O/Zx+jBrsAAHEoSqLgfaHGOpqo8DjwcnovC1fncF9y/YzPLtZfTvGc8jM8fzlbF9bbluYzrgtGG9iIoQlm4ttQThYUNawtDO8mp+v2gLb63dQ8+EaO69aDRXnTrQluo2phN6xEaRM7gny7aWcfcMp6MJDZYgwkj5oVoee287L6zYQWSEcOvZw5k9ZSjJcbZMtzH+MCW7Nw+8s5niyhr62Kg/SxDhoLqugaffz+fJZXkcrm/kazkDuHPaCPsPbIyfTc5O54F33P16X80Z4HQ4jrMEEcIaGl28kruLP/13G6UHazlvdB9+NH0Uw3v3cDo0Y7qk0VnJZCTFsmxbmSUILEGEJFVl4YZiHly4mbzSKnIG9eSvV0/gpEFpTodmTJcmIkwakc57m0todCmR3XzAhyWIELOyYB+/m7+J1TsPMLx3D/72rRymHdfbFtczJkia5kN8VniAEwf2dDocR1mCCBHbig/ywDub+e+mEvokx/LAFSdwxYT+RNnCesYE1aQRGYjAsq1lliCcDqC721txmIcXb+VfqwpJjIniR9NHct3pQ4iPsSGrxjghLTGGsf1SWLq1hDumjXA6HJ+oakBaGSxBOKTicD1P/O9z5n6Qjypcd8YQbj17OD0TY5wOzZhub0p2Bo8v2U5FdT0pCaE5jLy6roGlW0qZv76IksoaXrnpNL9fwxJEkNXUN/L8xzvc//kO13PZ+H5879xsBqQlOB2aMcZjysgMHn1vO8u3l3Hh2CynwzniYE09720uYcG6Iv63tYSaehc9E6I5b3Qm9Y0uv9d6sQQRJI0u5c01u/nDoq3sPnCYKdkZ/Hj6KEb3TXY6NGNMC+P6p5IUF8XSrSWOJ4gD1XUs2ljMO+uLWL6tjLpGFxlJsXz1pAHMGJPJxCFpAeurtAQRYKrKUs9iepuLDnJCvxQeunIspw+3tV6MCVVRkRFMGpHO0q2lAWvfb0vpwVoWbSzinfVFfPh5OY0upV9qPN88bRAzxmQyYWDPoKy5ZgkCmL9uL8f3TWZQr0S/fu5nhQf43fzNfJRXzsC0BB6bdSIXnpBli+kZEwamZGcwf10Rz3+8g149YoNyzaKKGt7ZUMTKgn2ouutU3DhpKDPGZDK2f0rQE1W3TxDVdQ3c+coa6hpcHJeVzIwxmcwYk8mIPh0vsFNQVsVDi7bw9md76ZUYwy8vPp5ZEwfaGvPGhJEp2b2JjhTueXNDUK+b3acHt00dwYwxmYzKTHJ0DpR0pTILOTk5mpub2+737dpXzcINRSxYX8SqHfsBGN67BzPGZDJ9TCajs5J9+kcqO1TLY+9u44UVO4mJiuDbk4Zy46QhJNliesaEpaKKGioO1wfteomxkfTvGdwBKyKySlVzvB6zBPFlxZU17mSxrogV+eW4FAamJRxJFuMHpB6VLKpqG3jq/XzmLPucmgYXsyYO4PZzRtA7yRbTM8aENksQHVR+qJbFG4tZsL6IDz8vo75RyUqJ4/zjM7nghCzGDUjh1dxCHvnvNsoO1XLBCZn84LyRDM2wxfSMMeHBEoQfVByu591N7mSxdGupu25tZAR1jS4mDknj7hmjuv20fGNM+GkrQXT7TmpfpcRHc/mE/lw+oT+HahtYsrmEj/PKOee43pw90hbTM8Z0PY4kCBGZDjwCRAJPqer9LY7HAn8HTgLKga+rakGw42xNj9govjKuL18Z19fpUIwxJmCCPu5SRCKBPwMzgNHALBEZ3eK0G4D9qjoceBh4ILhRGmOMcWJg/kRgu6rmqWod8DJwSYtzLgGe87z+F3COWBuOMcYElRMJoh+wq9l2oWef13NUtQGoAHp5+zARmS0iuSKSW1paGoBwjTGme3IiQXh7Emg5lMqXc9w7Veeoao6q5mRkZHQ6OGOMMW5OJIhCoHk18P7AntbOEZEoIAXYF5TojDHGAM4kiJXACBEZIiIxwEzgrRbnvAVc43l9JfCedqUJG8YYEwaCPsxVVRtE5FZgIe5hrs+o6gYR+RWQq6pvAU8D/xCR7bifHGYGO05jjOnuHJkHoarzgfkt9t3b7HUN8NVgx2WMMeYLXWqpDREpBXb44aPSgTI/fE5XYvfEO7svR7N7crRQvieDVNXrCJ8ulSD8RURyW1ubpLuye+Kd3Zej2T05WrjeE6tgY4wxxitLEMYYY7yyBOHdHKcDCEF2T7yz+3I0uydHC8t7Yn0QxhhjvLInCGOMMV5ZgjDGGONVt08QIjJARJaIyCYR2SAid3j2p4nIYhHZ5vmz29QTFZE4EflERNZ67skvPfuHiMgKzz15xbNUSrciIpEi8qmI/Mez3a3viYgUiMg6EVkjIrmefd32ewdARFJF5F8istnzc+W0cL0n3T5BAA3A91X1OOBU4BZPAaOfAO+q6gjgXc92d1ELTFXVccB4YLqInIq7cNPDnnuyH3dhp+7mDmBTs227J3C2qo5vNs6/O3/vgLta5juqOgoYh/v/S1jek26fIFR1r6qu9rw+iPsfsx9fLlr0HHCpMxEGn7od8mxGe74UmIq7gBN0s3sCICL9gQuBpzzbQje/J63ott87IpIMTMa9nhyqWqeqBwjTe9LtE0RzIjIYOBFYAfRR1b3gTiJAb+ciCz5PU8oaoARYDHwOHPAUcALvhZ66uj8BPwJcnu1e2D1RYJGIrBKR2Z593fl7ZyhQCsz1NEU+JSKJhOk9sQThISI9gNeAO1W10ul4nKaqjao6Hne9jonAcd5OC25UzhGRi4ASVV3VfLeXU7vNPfE4Q1Un4K4xf4uITHY6IIdFAROAJ1T1RKCKMGlO8sYSBCAi0biTwwuqOs+zu1hEsjzHs3D/Jt3teB6P/4e7fybVU8AJvBd66srOAC4WkQLcddSn4n6i6M73BFXd4/mzBHgd9y8T3fl7pxAoVNUVnu1/4U4YYXlPun2C8LQjPw1sUtU/NjvUvGjRNcCbwY7NKSKSISKpntfxwDTcfTNLcBdwgm52T1T1blXtr6qDcdcneU9Vr6Ib3xMRSRSRpKbXwHnAerrx946qFgG7RGSkZ9c5wEbC9J50+5nUInIm8D6wji/aln+Kux/iVWAgsBP4qqp2i7KnIjIWd0daJO5fIl5V1V+JyFDcvz2nAZ8CV6tqrXOROkNEzgJ+oKoXded74vm7v+7ZjAJeVNXfiEgvuun3DoCIjMc9kCEGyAOuw/N9RJjdk26fIIwxxnjX7ZuYjDHGeGcJwhhjjFeWIIwxxnhlCcIYY4xXliCMMcZ4ZQnCmA4SkT4i8qKI5HmWmvhIRC7zHJsoIstEZItnVc+nRCRBRK7wrJD7vmc4KCIyTERedvZvY8zRLEEY0wGeCZZvAMtUdaiqnoR7Al1/EekD/BP4saqOxL1MyTtAEvB93LPS/w58w/Nx9wH3BPmvYMwxRR37FGOMF1OBOlX9a9MOVd0BPCYivwKeU9WPPPsVz4qvIuICYoEEoFZEJgF7VXVbsP8CxhyLJQhjOuZ4YHUrx8bwxdLOLf0SWIh7zaarcc+unen36IzxA0sQxviBiPwZOBOoA3a1dp6qLsa9fDoicg0wHxgpIj/AXXDoDlWtDnzExhyb9UEY0zEbcK/SCYCq3oJ7YbYMz7GT2nqziCTgXrTtL8DvgOuBVcBVAYrXmHazBGFMx7wHxInIzc32JXj+fBy4RkROaTogIleLSGazc38EPKKq9UA87joSrmafYYzjrInJmA5QVRWRS4GHReRHuKuIVeEeuVQsIjOB34tIb9w/+JcB8wBEpC+Qo6q/8HzcH4CPgQOESSlK0z3Yaq7GGGO8siYmY4wxXlmCMMYY45UlCGOMMV5ZgjDGGOOVJQhjjDFeWYIwxhjjlSUIY4wxXv0/VgVkYamCqI0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "yt = []\n",
    "\n",
    "x = np.asarray(sorted(hist))\n",
    "for i in sorted(hist):\n",
    "    yt.append(hist[i])\n",
    "y = np.asarray(yt)\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.plot(x, y)\n",
    "\n",
    "ax.set(xlabel='GC%', ylabel='% of sequences')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "fig.savefig(\"test.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assignment\n",
    "\n",
    "A wig format is used for genome coverage of a dataset. The declaration line starts with the word variableStep and is followed by a specification for a chromosome. Ignore the optional span parameter for now. You have to write the code to read the \"80mM_Ctl_sub.wig\" file and then print out the coverage as follows:\n",
    "\n",
    "Chromosome Position Value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '_io.TextIOWrapper'>\n",
      "2L 21163512 105\n",
      "2L 21163513 97\n",
      "2L 21163514 97\n",
      "2L 21163515 97\n",
      "2L 21163516 97\n",
      "2L 21163517 95\n",
      "2L 21163518 94\n",
      "2L 21163519 79\n",
      "2L 21163520 75\n",
      "2L 21163521 74\n",
      "2L 21163522 73\n",
      "2L 21163523 72\n",
      "2L 21163524 72\n",
      "2L 21163525 72\n",
      "2L 21163526 71\n",
      "2L 21163527 68\n",
      "2L 21163528 68\n",
      "2L 21163529 68\n",
      "2L 21163530 68\n",
      "2L 21163531 68\n",
      "2L 21163532 68\n",
      "2L 21163533 68\n",
      "2L 21163534 68\n",
      "2L 21163535 68\n",
      "2L 21163536 68\n",
      "2L 21163537 68\n",
      "2L 21163538 68\n",
      "2L 21163539 66\n",
      "2L 21163540 55\n",
      "2L 21163541 53\n",
      "2L 21163542 53\n",
      "2L 21163543 53\n",
      "2L 21163544 53\n",
      "2L 21163545 53\n",
      "2L 21163546 50\n",
      "2L 21163547 50\n",
      "2L 21163548 50\n",
      "2L 21163549 50\n",
      "2L 21163550 50\n",
      "2L 21163551 18\n",
      "2L 21163552 9\n",
      "2L 21163553 9\n",
      "2L 21163554 9\n",
      "2L 21163555 9\n",
      "2L 21163556 9\n",
      "2L 21163557 9\n",
      "2L 21163558 9\n",
      "2L 21163559 9\n",
      "2L 21163560 4\n",
      "2L 21163561 3\n",
      "2L 21163562 3\n",
      "2L 21163563 4\n",
      "2L 21163564 7\n",
      "2L 21163565 7\n",
      "2L 21163566 7\n",
      "2L 21163567 7\n",
      "2L 21163568 7\n",
      "2L 21163569 7\n",
      "2L 21163570 7\n",
      "2L 21163571 7\n",
      "2L 21163572 7\n",
      "2L 21163573 11\n",
      "2L 21163574 17\n",
      "2L 21163575 17\n",
      "2L 21163576 17\n",
      "2L 21163577 17\n",
      "2L 21163578 16\n",
      "2L 21163579 16\n",
      "2L 21163580 16\n",
      "2L 21163581 16\n",
      "2L 21163582 16\n",
      "2L 21163583 16\n",
      "2L 21163584 16\n",
      "2L 21163585 16\n",
      "2L 21163586 16\n",
      "2L 21163587 17\n",
      "2L 21163588 26\n",
      "2L 21163589 32\n",
      "2L 21163590 35\n",
      "2L 21163591 35\n",
      "2L 21163592 36\n",
      "2L 21163593 36\n",
      "2L 21163594 35\n",
      "2L 21163595 65\n",
      "2L 21163596 65\n",
      "2L 21163597 65\n",
      "2L 21163598 65\n",
      "2L 21163599 70\n",
      "2L 21163600 70\n",
      "2L 21163601 70\n",
      "2L 21163602 75\n",
      "2L 21163603 75\n",
      "2L 21163604 75\n",
      "2L 21163605 76\n",
      "2L 21163606 76\n",
      "2L 21163607 76\n",
      "2L 21163608 69\n",
      "2L 21163609 70\n",
      "2L 21163610 69\n",
      "2L 21163611 69\n",
      "2L 21163612 69\n",
      "2L 21163613 69\n",
      "2L 21163614 75\n",
      "2L 21163615 75\n",
      "2L 21163616 110\n",
      "2L 21163617 109\n",
      "2L 21163618 109\n",
      "2L 21163619 123\n",
      "2L 21163620 123\n",
      "2L 21163621 123\n",
      "2L 21163622 119\n",
      "2L 21163623 122\n",
      "2L 21163624 122\n",
      "2L 21163625 121\n",
      "2L 21163626 119\n",
      "2L 21163627 120\n",
      "2L 21163628 120\n",
      "2L 21163629 120\n",
      "2L 21163630 120\n",
      "2L 21163631 120\n",
      "2L 21163632 115\n",
      "2L 21163633 96\n",
      "2L 21163634 93\n",
      "2L 21163635 94\n",
      "2L 21163636 93\n",
      "2L 21163637 93\n",
      "2L 21163638 94\n",
      "2L 21163639 93\n",
      "2L 21163640 96\n",
      "2L 21163641 95\n",
      "2L 21163642 92\n",
      "2L 21163643 92\n",
      "2L 21163644 91\n",
      "2L 21163645 92\n",
      "2L 21163646 87\n",
      "2L 21163647 83\n",
      "2L 21163648 84\n",
      "2L 21163649 82\n",
      "2L 21163650 68\n",
      "2L 21163651 36\n",
      "2L 21163652 34\n",
      "2L 21163653 35\n",
      "2L 21163654 34\n",
      "2L 21163655 34\n",
      "2L 21163656 34\n",
      "2L 21163657 22\n",
      "2L 21163658 21\n",
      "2L 21163659 18\n",
      "2L 21163660 18\n",
      "2L 21163661 17\n",
      "2L 21163662 21\n",
      "2L 21163663 20\n",
      "2L 21163664 20\n",
      "2L 21163665 21\n",
      "2L 21163666 22\n",
      "2L 21163667 22\n",
      "2L 21163668 22\n",
      "2L 21163669 23\n",
      "2L 21163670 22\n",
      "2L 21163671 20\n",
      "2L 21163672 17\n",
      "2L 21163673 17\n",
      "2L 21163674 17\n",
      "2L 21163675 17\n",
      "2L 21163676 17\n",
      "2L 21163677 17\n",
      "2L 21163678 17\n",
      "2L 21163679 17\n",
      "2L 21163680 18\n",
      "2L 21163681 18\n",
      "2L 21163682 19\n",
      "2L 21163683 35\n",
      "2L 21163684 36\n",
      "2L 21163685 36\n",
      "2L 21163686 38\n",
      "2L 21163687 39\n",
      "2L 21163688 39\n",
      "2L 21163689 35\n",
      "2L 21163690 35\n",
      "2L 21163691 41\n",
      "2L 21163692 41\n",
      "2L 21163693 42\n",
      "2L 21163694 42\n",
      "2L 21163695 42\n",
      "2L 21163696 42\n",
      "2L 21163697 42\n",
      "2L 21163698 42\n",
      "2L 21163699 42\n",
      "2L 21163700 42\n",
      "2L 21163701 42\n",
      "2L 21163702 42\n",
      "2L 21163703 46\n",
      "2L 21163704 55\n",
      "2L 21163705 55\n",
      "2L 21163706 57\n",
      "2L 21163707 58\n",
      "2L 21163708 58\n",
      "2L 21163709 58\n",
      "2L 21163710 58\n",
      "2L 21163711 59\n",
      "2L 21163712 59\n",
      "2L 21163713 59\n",
      "2L 21163714 59\n",
      "2L 21163715 59\n",
      "2L 21163716 59\n",
      "2L 21163717 59\n",
      "2L 21163718 60\n",
      "2L 21163719 60\n",
      "2L 21163720 60\n",
      "2L 21163721 61\n",
      "2L 21163722 63\n",
      "2L 21163723 63\n",
      "2L 21163724 63\n",
      "2L 21163725 63\n",
      "2L 21163726 63\n",
      "2L 21163727 63\n",
      "2L 21163728 63\n",
      "2L 21163729 63\n",
      "2L 21163730 59\n",
      "2L 21163731 58\n",
      "2L 21163732 61\n",
      "2L 21163733 64\n",
      "2L 21163734 64\n",
      "2L 21163735 64\n",
      "2L 21163736 64\n",
      "2L 21163737 64\n",
      "2L 21163738 64\n",
      "2L 21163739 64\n",
      "2L 21163740 64\n",
      "2L 21163741 64\n",
      "2L 21163742 64\n",
      "2L 21163743 64\n",
      "2L 21163744 64\n",
      "2L 21163745 64\n",
      "2L 21163746 64\n",
      "2L 21163747 64\n",
      "2L 21163748 64\n",
      "2L 21163749 64\n",
      "2L 21163750 64\n",
      "2L 21163751 67\n",
      "2L 21163752 69\n",
      "2L 21163753 69\n",
      "2L 21163754 69\n",
      "2L 21163755 69\n",
      "2L 21163756 69\n",
      "2L 21163757 69\n",
      "2L 21163758 69\n",
      "2L 21163759 68\n",
      "2L 21163760 68\n",
      "2L 21163761 67\n",
      "2L 21163762 67\n",
      "2L 21163763 67\n",
      "2L 21163764 67\n",
      "2L 21163765 66\n",
      "2L 21163766 66\n",
      "2L 21163767 66\n",
      "2L 21163768 66\n",
      "2L 21163769 66\n",
      "2L 21163770 66\n",
      "2L 21163771 66\n",
      "2L 21163772 66\n",
      "2L 21163773 66\n",
      "2L 21163774 67\n",
      "2L 21163775 67\n",
      "2L 21163776 67\n",
      "2L 21163777 67\n",
      "2L 21163778 68\n",
      "2L 21163779 68\n",
      "2L 21163780 68\n",
      "2L 21163781 68\n",
      "2L 21163782 67\n",
      "2L 21163783 67\n",
      "2L 21163784 67\n",
      "2L 21163785 66\n",
      "2L 21163786 65\n",
      "2L 21163787 65\n",
      "2L 21163788 64\n",
      "2L 21163789 63\n",
      "2L 21163790 63\n",
      "2L 21163791 63\n",
      "2L 21163792 63\n",
      "2L 21163793 63\n",
      "2L 21163794 62\n",
      "2L 21163795 62\n",
      "2L 21163796 62\n",
      "2L 21163797 62\n",
      "2L 21163798 62\n",
      "2L 21163799 62\n",
      "2L 21163800 61\n",
      "2L 21163801 61\n",
      "2L 21163802 61\n",
      "2L 21163803 60\n",
      "2L 21163804 60\n",
      "2L 21163805 60\n",
      "2L 21163806 58\n",
      "2L 21163807 55\n",
      "2L 21163808 56\n",
      "2L 21163809 56\n",
      "2L 21163810 57\n",
      "2L 21163811 60\n",
      "2L 21163812 61\n",
      "2L 21163813 61\n",
      "2L 21163814 61\n",
      "2L 21163815 61\n",
      "2L 21163816 61\n",
      "2L 21163817 62\n",
      "2L 21163818 63\n",
      "2L 21163819 63\n",
      "2L 21163820 62\n",
      "2L 21163821 62\n",
      "2L 21163822 60\n",
      "2L 21163823 58\n",
      "2L 21163824 59\n",
      "2L 21163825 59\n",
      "2L 21163826 59\n",
      "2L 21163827 59\n",
      "2L 21163828 59\n",
      "2L 21163829 59\n",
      "2L 21163830 59\n",
      "2L 21163831 58\n",
      "2L 21163832 57\n",
      "2L 21163833 57\n",
      "2L 21163834 57\n",
      "2L 21163835 54\n",
      "2L 21163836 41\n",
      "2L 21163837 38\n",
      "2L 21163838 29\n",
      "2L 21163839 28\n",
      "2L 21163840 27\n",
      "2L 21163841 27\n",
      "2L 21163842 29\n",
      "2L 21163843 29\n",
      "2L 21163844 29\n",
      "2L 21163845 31\n",
      "2L 21163846 34\n",
      "2L 21163847 34\n",
      "2L 21163848 34\n",
      "2L 21163849 28\n",
      "2L 21163850 30\n",
      "2L 21163851 34\n",
      "2L 21163852 33\n",
      "2L 21163853 33\n",
      "2L 21163854 33\n",
      "2L 21163855 28\n",
      "2L 21163856 28\n",
      "2L 21163857 26\n",
      "2L 21163858 25\n",
      "2L 21163859 25\n",
      "2L 21163860 25\n",
      "2L 21163861 25\n",
      "2L 21163862 26\n",
      "2L 21163863 35\n",
      "2L 21163864 35\n",
      "2L 21163865 34\n",
      "2L 21163866 34\n",
      "2L 21163867 34\n",
      "2L 21163868 33\n",
      "2L 21163869 33\n",
      "2L 21163870 33\n",
      "2L 21163871 35\n",
      "2L 21163872 38\n",
      "2L 21163873 37\n",
      "2L 21163874 37\n",
      "2L 21163875 37\n",
      "2L 21163876 39\n",
      "2L 21163877 39\n",
      "2L 21163878 39\n",
      "2L 21163879 39\n",
      "2L 21163880 38\n",
      "2L 21163881 38\n",
      "2L 21163882 38\n",
      "2L 21163883 39\n",
      "2L 21163884 39\n",
      "2L 21163885 40\n",
      "2L 21163886 40\n",
      "2L 21163887 42\n",
      "2L 21163888 42\n",
      "2L 21163889 42\n",
      "2L 21163890 43\n",
      "2L 21163891 43\n",
      "2L 21163892 43\n",
      "2L 21163893 43\n",
      "2L 21163894 43\n",
      "2L 21163895 43\n",
      "2L 21163896 44\n",
      "2L 21163897 47\n",
      "2L 21163898 47\n",
      "2L 21163899 47\n",
      "2L 21163900 47\n",
      "2L 21163901 41\n",
      "2L 21163902 37\n",
      "2L 21163903 37\n",
      "2L 21163904 37\n",
      "2L 21163905 38\n",
      "2L 21163906 38\n",
      "2L 21163907 38\n",
      "2L 21163908 34\n",
      "2L 21163909 34\n",
      "2L 21163910 34\n",
      "2L 21163911 34\n",
      "2L 21163912 34\n",
      "2L 21163913 34\n",
      "2L 21163914 35\n",
      "2L 21163915 35\n",
      "2L 21163916 35\n",
      "2L 21163917 32\n",
      "2L 21163918 32\n",
      "2L 21163919 32\n",
      "2L 21163920 32\n",
      "2L 21163921 32\n",
      "2L 21163922 32\n",
      "2L 21163923 32\n",
      "2L 21163924 32\n",
      "2L 21163925 32\n",
      "2L 21163926 34\n",
      "2L 21163927 34\n",
      "2L 21163928 33\n",
      "2L 21163929 33\n",
      "2L 21163930 33\n",
      "2L 21163931 32\n",
      "2L 21163932 32\n",
      "2L 21163933 32\n",
      "2L 21163934 32\n",
      "2L 21163935 32\n",
      "2L 21163936 32\n",
      "2L 21163937 32\n",
      "2L 21163938 32\n",
      "2L 21163939 32\n",
      "2L 21163940 32\n",
      "2L 21163941 32\n",
      "2L 21163942 32\n",
      "2L 21163943 33\n",
      "2L 21163944 34\n",
      "2L 21163945 35\n",
      "2L 21163946 35\n",
      "2L 21163947 35\n",
      "2L 21163948 36\n",
      "2L 21163949 38\n",
      "2L 21163950 38\n",
      "2L 21163951 37\n",
      "2L 21163952 38\n",
      "2L 21163953 40\n",
      "2L 21163954 40\n",
      "2L 21163955 40\n",
      "2L 21163956 40\n",
      "2L 21163957 41\n",
      "2L 21163958 41\n",
      "2L 21163959 42\n",
      "2L 21163960 42\n",
      "2L 21163961 42\n",
      "2L 21163962 42\n",
      "2L 21163963 42\n",
      "2L 21163964 42\n",
      "2L 21163965 42\n",
      "2L 21163966 41\n",
      "2L 21163967 41\n",
      "2L 21163968 41\n",
      "2L 21163969 41\n",
      "2L 21163970 41\n",
      "2L 21163971 41\n",
      "2L 21163972 42\n",
      "2L 21163973 41\n",
      "2L 21163974 43\n",
      "2L 21163975 43\n",
      "2L 21163976 43\n",
      "2L 21163977 44\n",
      "2L 21163978 45\n",
      "2L 21163979 45\n",
      "2L 21163980 45\n",
      "2L 21163981 45\n",
      "2L 21163982 42\n",
      "2L 21163983 42\n",
      "2L 21163984 43\n",
      "2L 21163985 43\n",
      "2L 21163986 44\n",
      "2L 21163987 45\n",
      "2L 21163988 43\n",
      "2L 21163989 43\n",
      "2L 21163990 44\n",
      "2L 21163991 44\n",
      "2L 21163992 44\n",
      "2L 21163993 44\n",
      "2L 21163994 42\n",
      "2L 21163995 44\n",
      "2L 21163996 45\n",
      "2L 21163997 45\n",
      "2L 21163998 45\n",
      "2L 21163999 48\n",
      "2L 21164000 48\n",
      "2L 21164001 54\n",
      "2L 21164002 54\n",
      "2L 21164003 54\n",
      "2L 21164004 54\n",
      "2L 21164005 54\n",
      "2L 21164006 54\n",
      "2L 21164007 54\n",
      "2L 21164008 54\n",
      "2L 21164009 54\n",
      "2L 21164010 55\n",
      "2L 21164011 55\n",
      "2L 21164012 55\n",
      "2L 21164013 55\n",
      "2L 21164014 55\n",
      "2L 21164015 55\n",
      "2L 21164016 55\n",
      "2L 21164017 51\n",
      "2L 21164018 54\n",
      "2L 21164019 63\n",
      "2L 21164020 62\n",
      "2L 21164021 61\n",
      "2L 21164022 63\n",
      "2L 21164023 70\n",
      "2L 21164024 68\n",
      "2L 21164025 65\n",
      "2L 21164026 64\n",
      "2L 21164027 72\n",
      "2L 21164028 74\n",
      "2L 21164029 74\n",
      "2L 21164030 65\n",
      "2L 21164031 64\n",
      "2L 21164032 64\n",
      "2L 21164033 66\n",
      "2L 21164034 62\n",
      "2L 21164035 58\n",
      "2L 21164036 82\n",
      "2L 21164037 108\n",
      "2L 21164038 108\n",
      "2L 21164039 107\n",
      "2L 21164040 107\n",
      "2L 21164041 109\n",
      "2L 21164042 111\n",
      "2L 21164043 111\n",
      "2L 21164044 111\n",
      "2L 21164045 126\n",
      "2L 21164046 156\n",
      "2L 21164047 156\n",
      "2L 21164048 155\n",
      "2L 21164049 158\n",
      "2L 21164050 159\n",
      "2L 21164051 159\n",
      "2L 21164052 162\n",
      "2L 21164053 163\n",
      "2L 21164054 163\n",
      "2L 21164055 164\n",
      "2L 21164056 164\n",
      "2L 21164057 164\n",
      "2L 21164058 164\n",
      "2L 21164059 166\n",
      "2L 21164060 166\n",
      "2L 21164061 168\n",
      "2L 21164062 168\n",
      "2L 21164063 166\n",
      "2L 21164064 168\n",
      "2L 21164065 176\n",
      "2L 21164066 176\n",
      "2L 21164067 181\n",
      "2L 21164068 180\n",
      "2L 21164069 180\n",
      "2L 21164070 181\n",
      "2L 21164071 180\n",
      "2L 21164072 182\n",
      "2L 21164073 200\n",
      "2L 21164074 200\n",
      "2L 21164075 200\n",
      "2L 21164076 201\n",
      "2L 21164077 203\n",
      "2L 21164078 205\n",
      "2L 21164079 207\n",
      "2L 21164080 210\n",
      "2L 21164081 210\n",
      "2L 21164082 210\n",
      "2L 21164083 209\n",
      "2L 21164084 209\n",
      "2L 21164085 209\n",
      "2L 21164086 209\n",
      "2L 21164087 209\n",
      "2L 21164088 212\n",
      "2L 21164089 210\n",
      "2L 21164090 210\n",
      "2L 21164091 210\n",
      "2L 21164092 210\n",
      "2L 21164093 210\n",
      "2L 21164094 210\n",
      "2L 21164095 210\n",
      "2L 21164096 210\n",
      "2L 21164097 211\n",
      "2L 21164098 211\n",
      "2L 21164099 211\n",
      "2L 21164100 212\n",
      "2L 21164101 212\n",
      "2L 21164102 212\n",
      "2L 21164103 209\n",
      "2L 21164104 208\n",
      "2L 21164105 212\n",
      "2L 21164106 212\n",
      "2L 21164107 212\n",
      "2L 21164108 212\n",
      "2L 21164109 212\n",
      "2L 21164110 213\n"
     ]
    }
   ],
   "source": [
    "wiggly_file = open('80mM_Ctl_sub.wig', 'r' )\n",
    "print(type(wiggly_file))\n",
    "\n",
    "cpv_list = [] #chrom-pos-value list\n",
    "\n",
    "for i in wiggly_file: \n",
    "    if i.startswith(\"var\"): \n",
    "        chrom = i.lstrip('variableStep chrom=').replace(\"span=1\", \"\")\n",
    "    else:\n",
    "        position_value = str(i)\n",
    "        position_value.replace(\"\\n\", \" \")\n",
    "        cpv_list.append(chrom + position_value)\n",
    "#print(cpv_list) #ugly, need to get rid of /n stuff\n",
    "\n",
    "for i in cpv_list:\n",
    "    fixed_item = i.replace(\"\\n\",\"\")\n",
    "    print(fixed_item)\n",
    "    \n",
    "wiggly_file.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the fastq file that we worked on, calculate the GC content at each position of the read. Here is a demo:\n",
    "\n",
    "`sequence 1: ATGCATGC\n",
    "sequence 2: GAGAGCTA`\n",
    "\n",
    "First position is 50% GC, second position is 0% GC and so on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq #1 Seq: GCTGGCTGCGTCTGTGGTGGGTTTCATGTTAAGGTC Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #2 Seq: GTGAAAATCAGGGAAAATGAGAAACATCCACTTGAC Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #3 Seq: GGAGTTTTTCTCTCCCTGATGCCAGAAGACTGCCTG Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #4 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #5 Seq: GTTTTCTTGCCATATTCCACGTTCTACAGAGGACAT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #6 Seq: GGAACTGGCTCCTTTTCCATCCCTGCCTGGGGTGGC Number of GC: 23 Sequence Length: 36 GC Content: 63.89\n",
      "\n",
      "Seq #7 Seq: GTGAAATATGGCAAGAAAACTGAAAATCATGGAAAA Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #8 Seq: GGAAAATTTAGAAATGTCCACTGTAGGACGTGGAAT Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #9 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #10 Seq: GAAAAATGACGAAATCACTGAAAATCGTGAAAAATG Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #11 Seq: GTTGGTTTGGAGAGGTAAGCACCATGTTTGAAAGTG Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #12 Seq: GTTTCCAAGCTGAGATCCTAGGTCTGGAAGATGAAA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #13 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #14 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #15 Seq: GTCAGGCTGCAACAGCATCGAGCTGTAACCTGTGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #16 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #17 Seq: GTCATTTTTCAAGTCGTCAAGTGGATGTTTCTCATT Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #18 Seq: GCAAAGATGAAATACAGACTTAGAAAGTGTTAACTC Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #19 Seq: GGAACATGAGAAATACACACTTTAGGACGTGAAATA Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #20 Seq: GTTTAATGGCTGCAAAGAGTTTGCCTCAATCTAGAA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #21 Seq: GTGAGGGGATCAAAGGAGGGCCGGGCACCGTCTTTT Number of GC: 22 Sequence Length: 36 GC Content: 61.11\n",
      "\n",
      "Seq #22 Seq: GGTTTGGGGTTTCCTTAAGTTTGTTATCAGCAAGAT Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #23 Seq: GAGTTTTGTTTCCTTCTGCAAACATTCCTGCATTCC Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #24 Seq: GCAATAGTGAATAAGAGACCCAGTCGCAGCGGAGGT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #25 Seq: GTGAAAAATATGAAATGCACACTGAAGGACCTGGAA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #26 Seq: GAAAAATGAGAAATGCACACTGAAGGACCTGGAATA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #27 Seq: GGGTATCCTATCCTGGCGGCAAGCACCACTGGCTTA Number of GC: 21 Sequence Length: 36 GC Content: 58.33\n",
      "\n",
      "Seq #28 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #29 Seq: GATATACCAGACCCTCCAAGTACTTCCTTACAGATC Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #30 Seq: GATCGGAAGGAGCTCGTATGCCGTCTTCTGCTTGGA Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #31 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #32 Seq: GTCCAGACTCACTTCAAACTCTTGATCCTTCAGACT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #33 Seq: GGATTATAGGCAAATGCTACCAAGCTTGGCTAGGGA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #34 Seq: GTATGGAATCGGCTACAGTAGGACGTAAACGAGCTT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #35 Seq: GTGAGTGGCCTGTACTTTCACCTAATTCCATGTCAA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #36 Seq: GGAAAAAAAGGCAAGACAGAAGCAGTATCTAGGACT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #37 Seq: GATCAGGGGGATACAAATTGGAAAAGAAGAAGTGAA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #38 Seq: GTTTTTCATGACTGGGTGCCACAAGACCACCCCTCT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #39 Seq: GCATTCAGCTGGCAACTTGAAATGCTACATTCCCCC Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #40 Seq: GTTTAGTAGTAAATTAGACAACAGGAAAACACAGAG Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #41 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #42 Seq: GACAAAAATGAGATTTCACTATTTAGCATGGACATT Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #43 Seq: GAGAGATAAGGAAAACAAACATGGGATAGATAAAGC Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #44 Seq: GCTATTTTCTTCCTGGGTCTAGCTCTACATTGGCCT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #45 Seq: GGTTTCGAAGCAGTGGTGTGAGATCGGCTTCCAAGG Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #46 Seq: GATCAAATTTCTCTCTCTCTTGTGTGTGTATATGGT Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #47 Seq: GAAGGTAATTGGGATCTATCCCATATGAGAAGCACT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #48 Seq: GAAAATGAGAAATACACACTTTAGGACGTGAAATAT Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #49 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #50 Seq: GCTGTTGTGTGTGGTGTTATAGTTAATTTGGGTGAT Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #51 Seq: GGACTAAGAGAGAAGTAATACTATAAACATGATATG Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #52 Seq: GGAAAAGTCCGAATGAGCAACATAGTAAATCACATA Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #53 Seq: GTCTTTTATCGTATGAATAACATTTTGTCTCCTGGG Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #54 Seq: GAAAGTGACAAAAGGGGTCTCTCTGATTTGAGAGCT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #55 Seq: GATGAGCTATGAGGTACTGATGTGGTGGCAAACCAT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #56 Seq: GTATTTGTGTACAGTATGCAGAGAGTAACTTCTTTC Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #57 Seq: GAAAATGTAGGGATGGTATGCGGGGGTGAGGTGAGC Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #58 Seq: GTTGGGTTTTTTTTTTTCCCTTCTTGGGGGTTGATC Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #59 Seq: GTTTTGCTTGGGTGAAGATCTGTGTGAAGAATGCAA Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #60 Seq: GAAATCCTTCTGAGGTTAGGGATCTAGGGGAAGTGG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #61 Seq: GAGAAACATCCACTTGACGACTTGAAAAATGATGAA Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #62 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #63 Seq: GTTTTGTAACCTGGGTCCAAACTGAAGGGCCCATTG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #64 Seq: GTGTGAGGCATCAGGACATAGAGAGTTGGGTTACAT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #65 Seq: GGGAGATATTATACAGACCATAAGTAAGACCATCCA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #66 Seq: GTTTTCCTCACCATATTTCACGTCCTAAAGTGTGTA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #67 Seq: GGAAAATGAGAAATACACACTTTAGGACGTGAAATA Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #68 Seq: GAAAAATGAGAAATACACACTTTAGTACGTGAAATA Number of GC: 10 Sequence Length: 36 GC Content: 27.78\n",
      "\n",
      "Seq #69 Seq: GTACCTAGTAGAGAACCTGAACACAGTAGAAACTAG Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #70 Seq: GAAGGAAAAGTCATCTAAAGGTCAGAGTCCACAGAC Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #71 Seq: GTTTTTAGTGTGCCGTGGCTTCACACCCTTACCAAG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #72 Seq: GTGCTCTTACCTGCTGAGCCATCTCACCAGCCAGAA Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #73 Seq: GATCGGAAGAGCTCGTTTGCCGTCTTCTGCTCAGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #74 Seq: GTTTATTTTCTAAGATAGAGTCTTACAGTGTAGCCC Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #75 Seq: GTCATTTTTCAAGTCGTCAAGTGGATGTTTCTCATT Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #76 Seq: GGAAATACAGTGACAAAGAAAGGGGCAGAGACTGAA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #77 Seq: GACAGGCTATGAGGGAGAAAGATACCAGTCAGATGA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #78 Seq: GTAGACACATCAAGAATCAGGGAAGAGGATCGGTTT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #79 Seq: GATCGGAAGAGGCTCGTATGCCGTCTTCTGCTTGGA Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #80 Seq: GTTTCTGCAGTAACCTTGAAGTTGGCTGGAGACAAG Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #81 Seq: GTATGTGGGGGAGAGAGATAAGGAAGGGAGAGACAG Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #82 Seq: GGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGATAG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #83 Seq: GATGGCTCAGTGGTTAAGAGCGCCAACTGCTCTTCC Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #84 Seq: GATCAAGTCTGGGAAGGTAGATCTCGTGTCACTGTA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #85 Seq: GTTTTTAGCTACAAAATAATCAAGAATGTGTTTAAT Number of GC: 8 Sequence Length: 36 GC Content: 22.22\n",
      "\n",
      "Seq #86 Seq: GCAGTGACAGAAGGAAGGGAGACACAGCCCTAGTTA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #87 Seq: GTTAGCATGAGAGCCTATACCGTGGACAGGTGAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #88 Seq: GTGTTTTTTCAATAAAGTTTTTATAGTCAGGCTCAT Number of GC: 10 Sequence Length: 36 GC Content: 27.78\n",
      "\n",
      "Seq #89 Seq: GTTGGTGGTTCTGTCTCTGTGAGTCCCCATATTCCA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #90 Seq: GCAGATCACCCATGCCTCAGAGTAAGAAGGCACACG Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #91 Seq: GAAGCTGTTTTCCTGCTACTCTCTGAGCCAGCTCCC Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #92 Seq: GTTTTCAAGTCGTCAAGTGGATGTTTCTCATTGTCC Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #93 Seq: GATTAAGAGTGACTTTCCTGGTGTTGCTGTCAGTGC Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #94 Seq: GAGCCTCACTTGCTCCTTCCTGGAAGGAAAGAAAGT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #95 Seq: GAATGTGTTTTTCAGTGTAACTCACTCATCTAATCT Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #96 Seq: GAAATCACTAAAAAACGTGAAGAATGAGAAATGCAC Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #97 Seq: GAGGATTTCTGAGTTCAAGGCCAGCCTGGTATATAG Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #98 Seq: GAAAACTGTTGAACCTTCCAGTCTGCTGTCCACATA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #99 Seq: GAAATGTGCACTGTAGGACGTGGAATATGGCAATAA Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #100 Seq: GTATTTAATTAAAAAAAGTTTTATGATATGCACATG Number of GC: 7 Sequence Length: 36 GC Content: 19.44\n",
      "\n",
      "Seq #101 Seq: GTCCACTGTAGGACGTGGAATATGGCAAGAAAACTG Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #102 Seq: GTGTACAACAACCCAAACCCTTATCCACATTCACTC Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #103 Seq: GGAAAGTGCTCACTACAGTTCACGAAGAACACCGGT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #104 Seq: GTAGCAAAGCCACCGTAGTTAGCTGGGAATGCAATG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #105 Seq: GAAAGTAAACTGCTGCCTTGCTCTGCTACATGCATT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #106 Seq: GAGATACAAAGGAAACAAAGTACAAACTAGTGGGTA Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #107 Seq: GTGGTCCAGCATCTGAGAGATCCTGAGGGCCCAGGC Number of GC: 23 Sequence Length: 36 GC Content: 63.89\n",
      "\n",
      "Seq #108 Seq: GTAGTTAATTGATGCTGGGTGGTGGCAGCACACACC Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #109 Seq: GATCGGAAGAAGCTCGTATGCCGTCTTCTGCTTGGA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #110 Seq: GAGGAAAAGTGAAAAAGTTGGAAAATTTAGAAATGT Number of GC: 10 Sequence Length: 36 GC Content: 27.78\n",
      "\n",
      "Seq #111 Seq: GATGGTGCTACTCCCTGGTTCAAGGATATACAAATC Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #112 Seq: GATGTTTCTCATTTTCCATGATTTTCAGTTTTCTTG Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #113 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #114 Seq: GAGGAAATATGTGGATGTGTTTACACTCTGAGGCAG Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #115 Seq: GTCTAGAACTGAAAAGCAAACTTGTATCATCTGATT Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #116 Seq: GAAACTGAAACCATGTATTATCTGTCCTCATTTATT Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #117 Seq: GATCGGAAGAAGCTCGTATGCCGTCTTCTGCTTGGA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #118 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #119 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #120 Seq: GAGGGGGACTGGAAACAGGTGAAATAAAGGGGGGAA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #121 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #122 Seq: GCTCGTGCATGAATACATGCCTGAAGTTAAATTGAA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #123 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGAAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #124 Seq: GGTGGGTCTGTGAGTTGAAGCCAGCCGGGTCTACAT Number of GC: 21 Sequence Length: 36 GC Content: 58.33\n",
      "\n",
      "Seq #125 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #126 Seq: GTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #127 Seq: GGTTTAATAGGGAATCAGACAGTGGGGACAAGCAGA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #128 Seq: GAGGAGAAGGCATCCATAGTAGGGTTGAGTGGCTTA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #129 Seq: GGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #130 Seq: GAAGCAAATCAAAGACCAACTTGGTCTGCATGAGAC Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #131 Seq: GTGAGTATGCTTGGAAAAACATTTTAGGAATTTCCA Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #132 Seq: GCTGATTATCTACACAAGTCACTCTTAAAGGGACCA Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #133 Seq: GAAATGGTTATTTCAAAAGCACCACAGCAGGTTCTG Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #134 Seq: GTGGGGGAAATTAGAACTAACACCTTTTAAAATTTA Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #135 Seq: GTGTATTTCTCATTTTCCGTGATTTTCAGTTTTCTC Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #136 Seq: GCCATTATGTGTACAACAAGGGAGGTCTACATAAGA Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #137 Seq: GATATTAATTATCTTGGGCCATTTTTAGCTCCCTTA Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #138 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #139 Seq: GAAATAAAGAAATTAAAGACTTTTAGAGTTTAATGA Number of GC: 7 Sequence Length: 36 GC Content: 19.44\n",
      "\n",
      "Seq #140 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #141 Seq: GCGAGAAAACTTAAAATCACGGAACATGAGAAATAC Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #142 Seq: GTAGCCATAGAAAAGTAACTTAGACGTCAGGGACGT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #143 Seq: GTCTATCTCACAAGAATCCTGGGTTGGAAGTAGGCC Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #144 Seq: GTTTGGGCTACAGAGAGTATTCAAGGCAGAGCTGAG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #145 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #146 Seq: GTGAAAAATGAGAAATGCCCACTGAAGGACCTGGAA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #147 Seq: GGAGGGAAGAATGAGAGGGGAGAGTCTGGAGAACAG Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #148 Seq: GTTCTTCAAGCTGTGCAAACGACCATGGTTAGATTC Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #149 Seq: GTTTGTGTTTCCCTCCCAAGGGAAGTACAGACGCCT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #150 Seq: GCTACAAGACAGGAATTGACAGATGTAAACTATCCA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #151 Seq: GAAAACTAAGGAGCCCATTCCCTCATGGCTCAGCAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #152 Seq: GAAAACCAGTGCATGATGGGTAAATAGTGTTCTCTT Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #153 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #154 Seq: GAAAATGCACACTGTAGGACCTGGAATATGGTGAGA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #155 Seq: GCTCACACTGGCCTTTTAGGTTTTCTCCTAATTCCA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #156 Seq: GATCGGAATAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #157 Seq: GCAAGGAATGCTGGTGCTCAGGTCACTAATTTCTTT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #158 Seq: GCTTTACCCAACCATCCCATCTTAGTTTTCATCGTA Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #159 Seq: GATGAAGAAGAGGCTGATGCCTGTGGTGCACAGGAA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #160 Seq: GAAGGAAGGTTTAACTGAGGAAGCAATGCAACTTTT Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #161 Seq: GCTACTATACCCAGACAAACTCTCAATTACCATAGA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #162 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #163 Seq: GTGGAAGCCATCACCATAGCGTTGTCCTCTGACCTC Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #164 Seq: GAAGCTGATTATGGGATGGATCCCTGGATATGGCAA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #165 Seq: GATAACCAGGTCTGCTACTGGCATATTCGACTCAAG Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #166 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #167 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #168 Seq: GTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #169 Seq: GACAGACCTGGAAAGGGCATCAGATCCCATTACAGA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #170 Seq: GGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #171 Seq: GGATGTTTCTCATTTTCCATGATTTTCAGTTTTCTT Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #172 Seq: GTGGTTGTACCAGTTTGCAATCCCACCAGCAGTGGA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #173 Seq: GAGTATAAGTCAGCAGGGGAGGTGGCTTCATTCTTG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #174 Seq: GTTCAAGGCCAGACTTACCTTGTAAAGCTATACTTA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #175 Seq: GTTTTTCAAGACGAGATTTCTCAGTGTAGCCCTGGC Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #176 Seq: GTTCACAAAATAAGTATACAGGCCTTTCCAAGGGTC Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #177 Seq: GGAAATACTAACAAACAGAAGAAAACTTCAACTCAA Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #178 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #179 Seq: GGTTAATTGTGTAGTTGCAGAATAAACACACAGAAC Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #180 Seq: GTGTGGAGGGATCTTGAACTTCTTGTTGTCTGGTCT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #181 Seq: GATTTTCAGATTTCTTGCCATATTCCACGTCCTACA Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #182 Seq: GCTCTGCTGAGGCTCTTCAAAGTGGGTGGGCAATTA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #183 Seq: GAAGTTGTTTCTGATGCTGAATGGGGAAAAGAGCAT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #184 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #185 Seq: GAATCATAGCTAGAATGATACTGACATGCACCCCAG Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #186 Seq: GGAGATGAGGAAGAGGTTCCGAATAGCTGAGCATCA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #187 Seq: GACATTTCTAAATTTTCCACTTTTTTCAGTTTTCCT Number of GC: 10 Sequence Length: 36 GC Content: 27.78\n",
      "\n",
      "Seq #188 Seq: GTTTTTCAGTGATTTCGTCATTTTTCAAGTCTTCAA Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #189 Seq: GGGAAAAATATCAGGGCACAGTCCCAATCAAAGAAA Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #190 Seq: GTCAGAGTGGCCTTCTTCATGCAGTATTCTCAAGCA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #191 Seq: GTGATGGAGTTTTGATGCTGCCGGTTTTCACCTTTT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #192 Seq: GAGGTGTAAGCTGAATAAACCCTTTCCTCCCCAACT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #193 Seq: GCTGTGGTAATGGGTGCTTATTCTGCATTCTGGCCA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #194 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #195 Seq: GGAATGGGGGAGTAGGTGCTGGTTAAGCTAACTAGA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #196 Seq: GGAAAATGAGAAATACACACTTTAGGACGTGAAATA Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #197 Seq: GTATGGGAGGGAAACAAGGAAAGAGTCTAATGAAGG Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #198 Seq: GGTTTATTGGCTGTCTTGGAACTCACTCTGTAGACC Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #199 Seq: GAGAAATACACAATTTAGGACGTGAAAATGGCGAGG Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #200 Seq: GTTTAGAAACCCTTTACTGTTGATAATTTTTTCTCT Number of GC: 10 Sequence Length: 36 GC Content: 27.78\n",
      "\n",
      "Seq #201 Seq: GCAAAGCCTCTTCACTTTCTCTCGCCCTGAATTACT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #202 Seq: GATCGGAAGCTCGTATGCCGTCTTCTGCTTGGAAAA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #203 Seq: GAGAAAATTGAAAATCACAGAAAAAGAGAAATACAC Number of GC: 10 Sequence Length: 36 GC Content: 27.78\n",
      "\n",
      "Seq #204 Seq: GAAAACATGATTTTAAGTAAAAGAAGACCAGCACAC Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #205 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #206 Seq: GGGATTAGAAGACATATGGCTGACCTATGACCCTTG Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #207 Seq: GAAACTCCGACTACAGTGTTCAGAAGAGGCAGAGAT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #208 Seq: GAAGTATGGCCTTGTTGGACTAAGTGTAGCCCTTTT Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #209 Seq: GGATGCCCTACATTTGTACTCTTTGGTTGGTAGTTT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #210 Seq: GAGACTATTGCCCATGATGGTGGAGTGGAAGCATGG Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #211 Seq: GAAGACTTCTCTTAGTTCTCAGGACATGTGATACTT Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #212 Seq: GATAGTGGATTTTGAATCTTGCTTCAAATATGCTTG Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #213 Seq: GAAAATATTTCCTGTGGCTTTCACCTGAGTTTCTTG Number of GC: 14 Sequence Length: 36 GC Content: 38.89\n",
      "\n",
      "Seq #214 Seq: GCTGGCATGGAGTCTTAGCTTATGGCCCTTGTGGAA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #215 Seq: GATGTCCAGGGCGTCAATAATAAAGAGAGAGCAGCG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #216 Seq: GGAAAATGATAAAAACCACACTGTAGAACATATTAG Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #217 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #218 Seq: GAGCAGAGTCAGAAACCTCAAGCCAGGCTGAAGAAC Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #219 Seq: GGGAAAAGAGTAATATGGTTCCTGCCAGTGGGACAA Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #220 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #221 Seq: GCTATAAAAGAATAAAGACATATAGGCTCTTAGTAC Number of GC: 11 Sequence Length: 36 GC Content: 30.56\n",
      "\n",
      "Seq #222 Seq: GAATCATTGCTTTAAGCACAAGAGGGGCAAACAGGA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #223 Seq: GAAGAAAGTCTGAGCTGGGCAGATCCTGGGAAGAAA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #224 Seq: GAAATCGTTAAATCAGTCCTGTGAGGTGAGCATGTT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #225 Seq: GACCAGTCATGCTGCTCAGTTGCTAACAGTGTGTGT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #226 Seq: GAAGCTTCAAGCCTGAGGGAAAGGGTACTAGGGACA Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #227 Seq: GTGATTTTCAGTTTTCTCGCCATATTCCAGGTCCTT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #228 Seq: GACAGACTTTCTCTATGTAGCCCTGACTGACCTATA Number of GC: 16 Sequence Length: 36 GC Content: 44.44\n",
      "\n",
      "Seq #229 Seq: GGAAAATGAGAAATACACACTTTAGGACGTGAAATA Number of GC: 12 Sequence Length: 36 GC Content: 33.33\n",
      "\n",
      "Seq #230 Seq: GATGTCCTGTAACATCAGCCCTGCTGCTTCACTTTG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #231 Seq: GAAAGAAATAAAGAAACAAAACCCTACCTGAAGCAG Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #232 Seq: GGATGAGGCCCTGTGATTCTCATCCCACAGTGAGAG Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #233 Seq: GGAAAGGAAAAGAAAGGAAAGGAAAGTGTGAATTGA Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #234 Seq: GTGTGTATGTAAGCCAAAAGAGGACATTAGGTGTCT Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #235 Seq: GTGAAGCAGTCAGGCAGGGACATGGCAGGGGTGCAC Number of GC: 23 Sequence Length: 36 GC Content: 63.89\n",
      "\n",
      "Seq #236 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #237 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #238 Seq: GTGGAATATGGCAAGAAAAGGGAAAATCATGGAAAA Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "Seq #239 Seq: GAGGGATGGGTATGTGCCCATGAGTGCAGCTGCCCA Number of GC: 22 Sequence Length: 36 GC Content: 61.11\n",
      "\n",
      "Seq #240 Seq: GGAGGTGTGCAGTTGTCTAAGACTATCATCACTGGG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #241 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGGAT Number of GC: 19 Sequence Length: 36 GC Content: 52.78\n",
      "\n",
      "Seq #242 Seq: GGGTAAATGCATACACAGGAATATTTTCCTACTGCC Number of GC: 15 Sequence Length: 36 GC Content: 41.67\n",
      "\n",
      "Seq #243 Seq: GGAAACTGGGCCAGCCAGAAGCTGGGCTTATCTTAC Number of GC: 20 Sequence Length: 36 GC Content: 55.56\n",
      "\n",
      "Seq #244 Seq: GATGTGGACAAGATTCCAGACAGGACCTGTGAACTC Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #245 Seq: GAAGAAGGTGAGACTTGGGTTTTTGTGGCCTCTGTT Number of GC: 17 Sequence Length: 36 GC Content: 47.22\n",
      "\n",
      "Seq #246 Seq: GAAAGGATGGACCATGTAGAGACTGCCATATCCAGG Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #247 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #248 Seq: GATCCAGCTGTGTACTTGTATGCATGTCCCTTGGCA Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #249 Seq: GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTAGAT Number of GC: 18 Sequence Length: 36 GC Content: 50.0\n",
      "\n",
      "Seq #250 Seq: GTCTGACGTTACGAATAAATGTTAGTAAGCAGATTG Number of GC: 13 Sequence Length: 36 GC Content: 36.11\n",
      "\n",
      "{1: 55.56, 2: 38.89, 3: 52.78, 4: 50.0, 5: 41.67, 6: 63.89, 7: 30.56, 8: 38.89, 9: 52.78, 10: 30.56, 11: 44.44, 12: 44.44, 13: 52.78, 14: 50.0, 15: 52.78, 16: 52.78, 17: 36.11, 18: 33.33, 19: 36.11, 20: 38.89, 21: 61.11, 22: 38.89, 23: 41.67, 24: 52.78, 25: 38.89, 26: 38.89, 27: 58.33, 28: 52.78, 29: 44.44, 30: 55.56, 31: 52.78, 32: 44.44, 33: 47.22, 34: 47.22, 35: 44.44, 36: 41.67, 37: 38.89, 38: 52.78, 39: 50.0, 40: 33.33, 41: 52.78, 42: 30.56, 43: 36.11, 44: 47.22, 45: 55.56, 46: 36.11, 47: 41.67, 48: 30.56, 49: 52.78, 50: 38.89, 51: 30.56, 52: 36.11, 53: 36.11, 54: 44.44, 55: 47.22, 56: 36.11, 57: 55.56, 58: 41.67, 59: 41.67, 60: 50.0, 61: 36.11, 62: 52.78, 63: 50.0, 64: 47.22, 65: 38.89, 66: 38.89, 67: 33.33, 68: 27.78, 69: 41.67, 70: 44.44, 71: 50.0, 72: 55.56, 73: 52.78, 74: 36.11, 75: 36.11, 76: 44.44, 77: 47.22, 78: 44.44, 79: 55.56, 80: 47.22, 81: 52.78, 82: 50.0, 83: 55.56, 84: 47.22, 85: 22.22, 86: 52.78, 87: 50.0, 88: 27.78, 89: 50.0, 90: 55.56, 91: 55.56, 92: 41.67, 93: 47.22, 94: 50.0, 95: 33.33, 96: 33.33, 97: 47.22, 98: 44.44, 99: 41.67, 100: 19.44, 101: 47.22, 102: 44.44, 103: 50.0, 104: 50.0, 105: 44.44, 106: 36.11, 107: 63.89, 108: 52.78, 109: 52.78, 110: 27.78, 111: 44.44, 112: 30.56, 113: 50.0, 114: 44.44, 115: 33.33, 116: 30.56, 117: 52.78, 118: 52.78, 119: 52.78, 120: 52.78, 121: 50.0, 122: 38.89, 123: 50.0, 124: 58.33, 125: 50.0, 126: 50.0, 127: 47.22, 128: 50.0, 129: 50.0, 130: 44.44, 131: 33.33, 132: 41.67, 133: 41.67, 134: 30.56, 135: 33.33, 136: 41.67, 137: 33.33, 138: 52.78, 139: 19.44, 140: 50.0, 141: 36.11, 142: 44.44, 143: 50.0, 144: 50.0, 145: 50.0, 146: 44.44, 147: 55.56, 148: 44.44, 149: 52.78, 150: 38.89, 151: 50.0, 152: 38.89, 153: 52.78, 154: 44.44, 155: 44.44, 156: 50.0, 157: 44.44, 158: 41.67, 159: 52.78, 160: 38.89, 161: 38.89, 162: 52.78, 163: 55.56, 164: 47.22, 165: 47.22, 166: 52.78, 167: 50.0, 168: 50.0, 169: 50.0, 170: 50.0, 171: 30.56, 172: 52.78, 173: 50.0, 174: 38.89, 175: 47.22, 176: 41.67, 177: 30.56, 178: 52.78, 179: 36.11, 180: 47.22, 181: 38.89, 182: 52.78, 183: 41.67, 184: 52.78, 185: 44.44, 186: 50.0, 187: 27.78, 188: 30.56, 189: 41.67, 190: 47.22, 191: 44.44, 192: 47.22, 193: 50.0, 194: 52.78, 195: 50.0, 196: 33.33, 197: 44.44, 198: 47.22, 199: 41.67, 200: 27.78, 201: 47.22, 202: 50.0, 203: 27.78, 204: 33.33, 205: 50.0, 206: 47.22, 207: 47.22, 208: 44.44, 209: 41.67, 210: 52.78, 211: 38.89, 212: 33.33, 213: 38.89, 214: 52.78, 215: 50.0, 216: 30.56, 217: 52.78, 218: 52.78, 219: 47.22, 220: 52.78, 221: 30.56, 222: 44.44, 223: 50.0, 224: 41.67, 225: 50.0, 226: 52.78, 227: 41.67, 228: 44.44, 229: 33.33, 230: 50.0, 231: 36.11, 232: 55.56, 233: 36.11, 234: 41.67, 235: 63.89, 236: 52.78, 237: 52.78, 238: 36.11, 239: 61.11, 240: 50.0, 241: 52.78, 242: 41.67, 243: 55.56, 244: 50.0, 245: 47.22, 246: 50.0, 247: 50.0, 248: 50.0, 249: 50.0, 250: 36.11}\n"
     ]
    }
   ],
   "source": [
    "#very confused what is being asked \"at each position\" \n",
    "\n",
    "fastq_file = open('example.fastq', 'r' )\n",
    "\n",
    "seqs={} # A dictionary to store sequences if needed later\n",
    "\n",
    "gc_content={} # A dictionary to store GC content if needed later\n",
    "\n",
    "line_no = 0 # Variable to count lines\n",
    "\n",
    "\n",
    "for line in fastq_file: \n",
    "    line_no+=1 \n",
    "    if (line_no-2)%4 == 0: \n",
    "        key = int((line_no-2)/4)+1 \n",
    "        line=line.rstrip()\n",
    "        seqs[key] = line \n",
    "        gc = line.count('G') + line.count('C')\n",
    "        gc_p=round(gc*100/len(line),2) \n",
    "        gc_content[key] = gc_p\n",
    "        write_str = \"Seq #\"+ str(key) + \" Seq: \" + line + \" Number of GC: \" + str(gc) +\" Sequence Length: \"+ str(len(line)) \\\n",
    "        + \" GC Content: \" + str(gc_p) + \"\\n\"  \n",
    "        print(write_str)\n",
    "print(gc_content)\n",
    "fastq_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read a uniprot text file \"HistoneH3_Uniprot.txt\". Write code to extract just the subcellular localization and the amino acid sequence from the uniprot file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Nucleus. Chromosome.\n",
      "\n",
      "     MARTKQTARK STGGKAPRKQ LATKAARKSA PATGGVKKPH RYRPGTVALR EIRRYQKSTE\n",
      "\n",
      "     MPKDIQLARR IRGERA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "hissy = open('HistoneH3_Uniprot.txt', 'r' )\n",
    "\n",
    "#for line in hissy:\n",
    "#    print(line)\n",
    "\n",
    "for line in hissy:\n",
    "    if line.startswith(\"CC   -!- SUBCELLULAR LOCATION:\"):\n",
    "        print(line.replace(\"CC   -!- SUBCELLULAR LOCATION:\", \"\"))\n",
    "    if line.startswith(\"     M\"): #ideally would find the line starting with SQ and then somehow \"get next line\" but assuming here that I can generalize with \"    M\" \n",
    "        print(line)\n",
    "        \n",
    "hissy.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the fasta file we worked with, calculate the net charge of the protein. The negatively charged residues are 'D' and  'E'. The output should look like (These are made up numbers):\n",
    "\n",
    "Sequence_name:HuR Net_charge:4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence name:>HuR\n",
      "\n",
      "Net charge: -10\n",
      "Sequence name:>HuR\n",
      "\n",
      "Net charge: -4\n",
      "Sequence name:>HuR\n",
      "\n",
      "Net charge: -9\n",
      "Sequence name:>HuR\n",
      "\n",
      "Net charge: -1\n",
      "Sequence name:>HuR\n",
      "\n",
      "Net charge: -6\n",
      "Sequence name:>HuR\n",
      "\n",
      "Net charge: -1\n",
      "Sequence name:>Histone_H3\n",
      "\n",
      "Net charge: -2\n",
      "Sequence name:>Histone_H3\n",
      "\n",
      "Net charge: -7\n",
      "Sequence name:>Histone_H3\n",
      "\n",
      "Net charge: -2\n",
      "Sequence name:>Gapdh\n",
      "\n",
      "Net charge: -4\n",
      "Sequence name:>Gapdh\n",
      "\n",
      "Net charge: -7\n",
      "Sequence name:>Gapdh\n",
      "\n",
      "Net charge: -5\n",
      "Sequence name:>Gapdh\n",
      "\n",
      "Net charge: -3\n",
      "Sequence name:>Gapdh\n",
      "\n",
      "Net charge: -9\n",
      "Sequence name:>Gapdh\n",
      "\n",
      "Net charge: -5\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -9\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -9\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -5\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -11\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -9\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -3\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: -3\n",
      "Sequence name:>Actin\n",
      "\n",
      "Net charge: 0\n"
     ]
    }
   ],
   "source": [
    "fasta_file = open('protein_sequences.fasta', 'r' )\n",
    "\n",
    "\n",
    "for line in fasta_file:\n",
    "    if line.startswith('>'): \n",
    "        seqs_name = line\n",
    "    else:\n",
    "        seq = line\n",
    "        poscharges = seq.count('K') + seq.count('R') + seq.count('H')\n",
    "        countneg = seq.count('D') + seq.count('E')\n",
    "        netcharge = countpos - countneg\n",
    "        print(\"Sequence name:\"+seqs_name+\"\\nNet charge: \"+str(netcharge))\n",
    "fasta_file.close()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
